In [1]:
# data manipulation 
import math 
import numpy as np 
import pandas as pd 

# data visualization 
import matplotlib.pyplot as plt 

# data preprocessing 
from sklearn.preprocessing import MinMaxScaler # to normalize the price data 

# Modeling 
import tensorflow as tf 
from keras.models import Sequential # Deep learning library, used for neural networks
from keras.layers import SimpleRNN, LSTM, Dense # Deep learning classes for recurrent and regular densely-connected layers
from tensorflow.keras.optimizers import Adam, SGD 
from keras.wrappers.scikit_learn import KerasRegressor 
from sklearn.model_selection import RandomizedSearchCV
from keras.callbacks import EarlyStopping

# Data evaluation 
from sklearn.metrics import mean_squared_error 

### Import Dataset

In [2]:
data = pd.read_csv('../dataset.csv')
data.index = pd.to_datetime(data.index)
print(data.shape)
data.head()

(551, 11)


CPIAUCSL_PC1  UNRATE  FEDFUNDS    T10Y2Y        High         Low  \
1976-06-01       5.98131     7.6      5.48  0.801364  102.507728  100.982727   
1976-07-01       5.55556     7.8      5.31  0.981905  104.933810  103.501905   
1976-08-01       5.71956     7.8      5.29  1.137273  103.985909  102.575001   
1976-09-01       5.49451     7.6      5.25  1.178095  106.130951  104.637618   
1976-10-01       5.46448     7.7      5.02  1.427000  102.681905  101.036191   

            SP500 Close        Volume  ADX  RSI  BB width  
1976-06-01   101.766818  1.896500e+07  NaN  NaN       NaN  
1976-07-01   104.200953  1.875048e+07  NaN  NaN       NaN  
1976-08-01   103.291818  1.575818e+07  NaN  NaN       NaN  
1976-09-01   105.453333  1.975000e+07  NaN  NaN       NaN  
1976-10-01   101.885238  1.720905e+07  NaN  NaN       NaN

In [3]:
data.tail()

CPIAUCSL_PC1  UNRATE  FEDFUNDS    T10Y2Y         High  \
2021-12-01       7.09654     3.9      0.08  0.785000  4701.581854   
2022-01-01       7.52593     4.0      0.08  0.783500  4619.576025   
2022-02-01       7.91202     3.8      0.08  0.498947  4473.607345   
2022-03-01       8.55759     3.6      0.20  0.218261  4424.881645   
2022-04-01           NaN     NaN       NaN  0.155556  4500.185872   

                    Low  SP500 Close        Volume        ADX        RSI  \
2021-12-01  4642.169611  4674.772772  3.122720e+09  45.523755  85.192588   
2022-01-01  4528.042017  4573.815454  3.663972e+09  46.122325  78.669780   
2022-02-01  4392.072625  4435.980494  3.850936e+09  45.583801  70.709690   
2022-03-01  4351.570015  4391.265264  4.390362e+09  44.777416  68.295438   
2022-04-01  4439.167440  4468.859945  3.565105e+09  44.302466  70.197059   

            BB width  
2021-12-01  0.543355  
2022-01-01  0.501704  
2022-02-01  0.460878  
2022-03-01  0.419147  
2022-04-01  0.389122

### Data Preprocessing

In [6]:
# Feature Selection - Only Close Data
df = data.filter(['SP500 Close'])
data_unscaled = df.shift(-1).values
# print(len(data_unscaled))
# data_unscaled

# train_data_length = math.ceil(len(data_unscaled) * 0.8)
# print(train_data_length)

# Standardization using MinMaxScaler (Scaling data to a range between 0 and 1)
mmscaler = MinMaxScaler(feature_range=(0, 1))
data_scaled = mmscaler.fit_transform(data_unscaled)

In [7]:
# Set the sequence length - this is the timeframe used to make a single prediction
sequence_length = 50

# Prediction Index
index_Close = df.columns.get_loc('SP500 Close')
print(index_Close)

0


In [8]:
# Split the data into training and testing data sets
# As a first step, we get the number of rows to train the model on 80% of the data 
train_data_len = math.ceil(data_scaled.shape[0] * 0.8)
valid_data_len = math.ceil(train_data_len * 0.2)
print(train_data_len, valid_data_len)

# Create the training, validation, and testing data
train_data = data_scaled[0:train_data_len - valid_data_len, :]
valid_data = data_scaled[train_data_len - valid_data_len:train_data_len,:]
test_data = data_scaled[train_data_len:, :]
print(train_data.shape)
print(valid_data.shape)
print(test_data.shape)

441 89
(352, 1)
(89, 1)
(110, 1)


In [9]:
train_index = df.index[0:train_data_len - valid_data_len]
valid_index = df.index[train_data_len - valid_data_len:train_data_len]
test_index = df.index[train_data_len:]
print(len(train_index))
print(len(valid_index))
print(len(test_index))

352
89
110


If we have an N-dimensional space and evolve over t-time steps, our input representation must be of shape `(num_samples, t, N)`. In other words, the RNN needs data with the format of [samples, time steps, features]. Here, we are building an univariate model on the S&P 500. Using the monthly aggregation, we have 551 observations and only 1 feature with the sequence length initailized to be 50.  

In [10]:
def partition_dataset(sequence_length, df):
    x, y = [], []
    data_len = df.shape[0]
    for i in range(sequence_length, data_len):
        # x has (data_len - sequence_length) number of sequence_length data window 
        x.append(df[i-sequence_length:i,:])  
        # y has the prediction values for validation,  for single-step prediction
        y.append(df[i, index_Close]) 
    
    # Convert the x and y to numpy arrays
    x = np.array(x)
    y = np.array(y)
    return x, y

In [11]:
# Generate training data, validation, and test data
x_train, y_train = partition_dataset(sequence_length, train_data)
x_valid, y_valid = partition_dataset(sequence_length, valid_data)
x_test, y_test = partition_dataset(sequence_length, test_data)

# Print the shapes: the result is: (rows, training_sequence, features) (prediction value, )
print(x_train.shape, y_train.shape)
print(x_valid.shape, y_valid.shape)
print(x_test.shape, y_test.shape)

# Validate that the prediction value and the input match up
# The last close price of the second input sample should equal the first prediction value
print(x_test[1][sequence_length-1][index_Close])
print(y_test[0])

(302, 50, 1) (302,)
(39, 50, 1) (39,)
(60, 50, 1) (60,)
0.511380249717515
0.511380249717515


### Find Model Architecture

In [12]:
# Model with sequence_length Neurons 
neurons = sequence_length

# input shape = sequence_length Timestamps

def single_layer_RNN_model(optimizer='adam', activation='relu', learning_rate=0.01):
    model = Sequential()
    model.add(SimpleRNN(neurons, return_sequences=False, input_shape=(x_train.shape[1], 1)))
    model.add(Dense(1))
    if optimizer == 'adam':
        optimizer = Adam(learning_rate=learning_rate)
    elif optimizer == 'sgd':
        optimizer = SGD(learning_rate=learning_rate)
    model.compile(optimizer=optimizer, loss='mean_squared_error', metrics=[tf.keras.metrics.RootMeanSquaredError()])
    return model

def single_layer_LSTM_model(optimizer='adam', activation='relu', learning_rate=0.01):
    model = Sequential()
    model.add(LSTM(neurons, return_sequences=False, input_shape=(x_train.shape[1], 1)))
    model.add(Dense(1))
    if optimizer == 'adam':
        optimizer = Adam(learning_rate=learning_rate)
    elif optimizer == 'sgd':
        optimizer = SGD(learning_rate=learning_rate)
    model.compile(optimizer=optimizer, loss='mean_squared_error', metrics=[tf.keras.metrics.RootMeanSquaredError()])
    return model

def multiple_layer_LSTM_model(optimizer='adam', activation='relu', learning_rate=0.01):
    model = Sequential()
    model.add(LSTM(neurons, return_sequences=True, input_shape=(x_train.shape[1], 1))) 
    model.add(LSTM(neurons, return_sequences=False)) 
    model.add(Dense(1))
    if optimizer == 'adam':
        optimizer = Adam(learning_rate=learning_rate)
    elif optimizer == 'sgd':
        optimizer = SGD(learning_rate=learning_rate)
    model.compile(optimizer=optimizer, loss='mean_squared_error', metrics=[tf.keras.metrics.RootMeanSquaredError()])
    return model
    
def multiple_layer_LSTM_model_plus(optimizer='adam', activation='relu', learning_rate=0.01):
    model = Sequential()
    model.add(LSTM(neurons, return_sequences=True, input_shape=(x_train.shape[1], 1))) 
    model.add(LSTM(neurons, return_sequences=False))
    model.add(Dense(25, activation=activation))
    model.add(Dense(1))
    if optimizer == 'adam':
        optimizer = Adam(learning_rate=learning_rate)
    elif optimizer == 'sgd':
        optimizer = SGD(learning_rate=learning_rate)
    model.compile(optimizer=optimizer, loss='mean_squared_error', metrics=[tf.keras.metrics.RootMeanSquaredError()])
    return model

### Training the model

In [13]:
rnn_model = KerasRegressor(single_layer_RNN_model) 
params = dict(optimizer=['sgd', 'adam'], epochs=[50, 100], batch_size=[16, 32], activation=['relu','tanh'],
             learning_rate=[0.01, 0.001, 0.0005])
random_search = RandomizedSearchCV(rnn_model, params, cv=3)
rnn_model_results = random_search.fit(x_train, y_train, validation_data=(x_valid, y_valid),
                                          callbacks=[EarlyStopping(patience=3)])
print(f"Best: {rnn_model_results.best_score_} using {rnn_model_results.best_params_}")

<ipython-input-13-e209a15f0dae>:1: DeprecationWarning: KerasRegressor is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  rnn_model = KerasRegressor(single_layer_RNN_model)


Epoch 1/50
7/7 [==============================] - 1s 32ms/step - loss: 0.0134 - root_mean_squared_error: 0.1158 - val_loss: 0.0010 - val_root_mean_squared_error: 0.0317
Epoch 2/50
7/7 [==============================] - 0s 9ms/step - loss: 0.0020 - root_mean_squared_error: 0.0446 - val_loss: 8.0995e-04 - val_root_mean_squared_error: 0.0285
Epoch 3/50
7/7 [==============================] - 0s 8ms/step - loss: 0.0014 - root_mean_squared_error: 0.0380 - val_loss: 9.3765e-04 - val_root_mean_squared_error: 0.0306
Epoch 4/50
7/7 [==============================] - 0s 8ms/step - loss: 9.1375e-04 - root_mean_squared_error: 0.0302 - val_loss: 7.7868e-04 - val_root_mean_squared_error: 0.0279
Epoch 5/50
7/7 [==============================] - 0s 8ms/step - loss: 7.2986e-04 - root_mean_squared_error: 0.0270 - val_loss: 6.8410e-04 - val_root_mean_squared_error: 0.0262
Epoch 6/50
7/7 [==============================] - 0s 8ms/step - loss: 5.4666e-04 - root_mean_squared_error: 0.0234 - val_loss: 8.1841e-

7/7 [==============================] - 0s 7ms/step - loss: 4.0646e-05 - root_mean_squared_error: 0.0064 - val_loss: 6.9489e-04 - val_root_mean_squared_error: 0.0264
Epoch 29/50
7/7 [==============================] - 0s 7ms/step - loss: 3.9462e-05 - root_mean_squared_error: 0.0063 - val_loss: 6.1210e-04 - val_root_mean_squared_error: 0.0247
Epoch 30/50
7/7 [==============================] - 0s 7ms/step - loss: 3.9463e-05 - root_mean_squared_error: 0.0063 - val_loss: 6.6273e-04 - val_root_mean_squared_error: 0.0257
Epoch 31/50
7/7 [==============================] - 0s 7ms/step - loss: 3.9198e-05 - root_mean_squared_error: 0.0063 - val_loss: 6.5970e-04 - val_root_mean_squared_error: 0.0257
Epoch 32/50
7/7 [==============================] - 0s 7ms/step - loss: 3.7117e-05 - root_mean_squared_error: 0.0061 - val_loss: 5.7740e-04 - val_root_mean_squared_error: 0.0240
Epoch 33/50
7/7 [==============================] - 0s 7ms/step - loss: 3.7448e-05 - root_mean_squared_error: 0.0061 - val_loss:

7/7 [==============================] - 0s 7ms/step - loss: 0.0012 - root_mean_squared_error: 0.0351 - val_loss: 0.0025 - val_root_mean_squared_error: 0.0495
Epoch 21/100
7/7 [==============================] - 0s 6ms/step - loss: 0.0011 - root_mean_squared_error: 0.0333 - val_loss: 0.0021 - val_root_mean_squared_error: 0.0460
Epoch 22/100
7/7 [==============================] - 0s 7ms/step - loss: 0.0010 - root_mean_squared_error: 0.0318 - val_loss: 0.0019 - val_root_mean_squared_error: 0.0440
Epoch 23/100
7/7 [==============================] - 0s 6ms/step - loss: 9.5655e-04 - root_mean_squared_error: 0.0309 - val_loss: 0.0018 - val_root_mean_squared_error: 0.0419
Epoch 24/100
7/7 [==============================] - 0s 7ms/step - loss: 9.1088e-04 - root_mean_squared_error: 0.0302 - val_loss: 0.0016 - val_root_mean_squared_error: 0.0403
Epoch 25/100
7/7 [==============================] - 0s 7ms/step - loss: 8.7748e-04 - root_mean_squared_error: 0.0296 - val_loss: 0.0015 - val_root_mean_squ

Epoch 22/100
7/7 [==============================] - 0s 7ms/step - loss: 1.6034e-04 - root_mean_squared_error: 0.0127 - val_loss: 8.7121e-04 - val_root_mean_squared_error: 0.0295
Epoch 23/100
7/7 [==============================] - 0s 7ms/step - loss: 1.4575e-04 - root_mean_squared_error: 0.0121 - val_loss: 8.1954e-04 - val_root_mean_squared_error: 0.0286
Epoch 24/100
7/7 [==============================] - 0s 7ms/step - loss: 1.3425e-04 - root_mean_squared_error: 0.0116 - val_loss: 7.6560e-04 - val_root_mean_squared_error: 0.0277
Epoch 25/100
7/7 [==============================] - 0s 7ms/step - loss: 1.2373e-04 - root_mean_squared_error: 0.0111 - val_loss: 7.1957e-04 - val_root_mean_squared_error: 0.0268
Epoch 26/100
7/7 [==============================] - 0s 6ms/step - loss: 1.1400e-04 - root_mean_squared_error: 0.0107 - val_loss: 6.5535e-04 - val_root_mean_squared_error: 0.0256
Epoch 27/100
7/7 [==============================] - 0s 7ms/step - loss: 1.0798e-04 - root_mean_squared_error: 

13/13 [==============================] - 0s 5ms/step - loss: 7.4047e-04 - root_mean_squared_error: 0.0272 - val_loss: 0.0014 - val_root_mean_squared_error: 0.0376
Epoch 33/50
13/13 [==============================] - 0s 6ms/step - loss: 7.3231e-04 - root_mean_squared_error: 0.0271 - val_loss: 0.0014 - val_root_mean_squared_error: 0.0375
Epoch 34/50
13/13 [==============================] - 0s 5ms/step - loss: 7.2667e-04 - root_mean_squared_error: 0.0270 - val_loss: 0.0014 - val_root_mean_squared_error: 0.0374
Epoch 35/50
13/13 [==============================] - 0s 5ms/step - loss: 7.2073e-04 - root_mean_squared_error: 0.0268 - val_loss: 0.0014 - val_root_mean_squared_error: 0.0373
Epoch 36/50
13/13 [==============================] - 0s 5ms/step - loss: 7.1388e-04 - root_mean_squared_error: 0.0267 - val_loss: 0.0014 - val_root_mean_squared_error: 0.0372
Epoch 37/50
13/13 [==============================] - 0s 5ms/step - loss: 7.0968e-04 - root_mean_squared_error: 0.0266 - val_loss: 0.0014 

Epoch 29/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0013 - root_mean_squared_error: 0.0358 - val_loss: 0.0024 - val_root_mean_squared_error: 0.0491
Epoch 30/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0012 - root_mean_squared_error: 0.0352 - val_loss: 0.0023 - val_root_mean_squared_error: 0.0481
Epoch 31/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0012 - root_mean_squared_error: 0.0348 - val_loss: 0.0022 - val_root_mean_squared_error: 0.0471
Epoch 32/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0012 - root_mean_squared_error: 0.0344 - val_loss: 0.0021 - val_root_mean_squared_error: 0.0463
Epoch 33/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0012 - root_mean_squared_error: 0.0341 - val_loss: 0.0021 - val_root_mean_squared_error: 0.0455
Epoch 34/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0011 - root_mean_squared_error: 0.0338 - val_loss: 0.0020 - val_root_m

13/13 [==============================] - 0s 5ms/step - loss: 0.0052 - root_mean_squared_error: 0.0720 - val_loss: 0.0126 - val_root_mean_squared_error: 0.1121
Epoch 3/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0041 - root_mean_squared_error: 0.0641 - val_loss: 0.0101 - val_root_mean_squared_error: 0.1004
Epoch 4/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0035 - root_mean_squared_error: 0.0591 - val_loss: 0.0085 - val_root_mean_squared_error: 0.0921
Epoch 5/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0031 - root_mean_squared_error: 0.0558 - val_loss: 0.0074 - val_root_mean_squared_error: 0.0860
Epoch 6/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0029 - root_mean_squared_error: 0.0536 - val_loss: 0.0066 - val_root_mean_squared_error: 0.0815
Epoch 7/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0027 - root_mean_squared_error: 0.0520 - val_loss: 0.0061 - val_root_mean_squared_error

7/7 [==============================] - 0s 2ms/step - loss: 5.4306e-04 - root_mean_squared_error: 0.0233
Epoch 1/50
13/13 [==============================] - 1s 16ms/step - loss: 0.0032 - root_mean_squared_error: 0.0567 - val_loss: 0.0782 - val_root_mean_squared_error: 0.2797
Epoch 2/50
13/13 [==============================] - 0s 5ms/step - loss: 0.0022 - root_mean_squared_error: 0.0465 - val_loss: 0.0741 - val_root_mean_squared_error: 0.2722
Epoch 3/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0021 - root_mean_squared_error: 0.0454 - val_loss: 0.0714 - val_root_mean_squared_error: 0.2673
Epoch 4/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0020 - root_mean_squared_error: 0.0443 - val_loss: 0.0684 - val_root_mean_squared_error: 0.2615
Epoch 5/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0019 - root_mean_squared_error: 0.0433 - val_loss: 0.0677 - val_root_mean_squared_error: 0.2601
Epoch 6/50
13/13 [===========================

Epoch 47/50
13/13 [==============================] - 0s 6ms/step - loss: 2.5508e-04 - root_mean_squared_error: 0.0160 - val_loss: 0.0159 - val_root_mean_squared_error: 0.1259
Epoch 48/50
13/13 [==============================] - 0s 8ms/step - loss: 2.4388e-04 - root_mean_squared_error: 0.0156 - val_loss: 0.0151 - val_root_mean_squared_error: 0.1231
Epoch 49/50
13/13 [==============================] - 0s 8ms/step - loss: 2.3357e-04 - root_mean_squared_error: 0.0153 - val_loss: 0.0152 - val_root_mean_squared_error: 0.1233
Epoch 50/50
7/7 [==============================] - 0s 2ms/step - loss: 0.0095 - root_mean_squared_error: 0.0976
Epoch 1/50
13/13 [==============================] - 1s 19ms/step - loss: 0.0449 - root_mean_squared_error: 0.2119 - val_loss: 0.0817 - val_root_mean_squared_error: 0.2857
Epoch 2/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0257 - root_mean_squared_error: 0.1604 - val_loss: 0.0548 - val_root_mean_squared_error: 0.2340
Epoch 3/50
13/13 [====

13/13 [==============================] - 0s 5ms/step - loss: 0.0010 - root_mean_squared_error: 0.0323 - val_loss: 0.0019 - val_root_mean_squared_error: 0.0431
Epoch 45/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0010 - root_mean_squared_error: 0.0322 - val_loss: 0.0018 - val_root_mean_squared_error: 0.0430
Epoch 46/50
13/13 [==============================] - 0s 5ms/step - loss: 0.0010 - root_mean_squared_error: 0.0322 - val_loss: 0.0018 - val_root_mean_squared_error: 0.0427
Epoch 47/50
13/13 [==============================] - 0s 5ms/step - loss: 0.0010 - root_mean_squared_error: 0.0321 - val_loss: 0.0018 - val_root_mean_squared_error: 0.0425
Epoch 48/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0010 - root_mean_squared_error: 0.0321 - val_loss: 0.0018 - val_root_mean_squared_error: 0.0423
Epoch 49/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0010 - root_mean_squared_error: 0.0321 - val_loss: 0.0018 - val_root_mean_squared_

13/13 [==============================] - 0s 6ms/step - loss: 0.0010 - root_mean_squared_error: 0.0319 - val_loss: 0.0015 - val_root_mean_squared_error: 0.0388
Epoch 42/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0010 - root_mean_squared_error: 0.0318 - val_loss: 0.0015 - val_root_mean_squared_error: 0.0386
Epoch 43/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0010 - root_mean_squared_error: 0.0318 - val_loss: 0.0015 - val_root_mean_squared_error: 0.0385
Epoch 44/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0010 - root_mean_squared_error: 0.0317 - val_loss: 0.0015 - val_root_mean_squared_error: 0.0384
Epoch 45/50
13/13 [==============================] - 0s 6ms/step - loss: 0.0010 - root_mean_squared_error: 0.0317 - val_loss: 0.0015 - val_root_mean_squared_error: 0.0382
Epoch 46/50
13/13 [==============================] - 0s 6ms/step - loss: 9.9944e-04 - root_mean_squared_error: 0.0316 - val_loss: 0.0015 - val_root_mean_squa

Epoch 10/100
7/7 [==============================] - 0s 8ms/step - loss: 0.0095 - root_mean_squared_error: 0.0973 - val_loss: 0.0171 - val_root_mean_squared_error: 0.1306
Epoch 11/100
7/7 [==============================] - 0s 8ms/step - loss: 0.0084 - root_mean_squared_error: 0.0916 - val_loss: 0.0150 - val_root_mean_squared_error: 0.1225
Epoch 12/100
7/7 [==============================] - 0s 8ms/step - loss: 0.0075 - root_mean_squared_error: 0.0868 - val_loss: 0.0135 - val_root_mean_squared_error: 0.1163
Epoch 13/100
7/7 [==============================] - 0s 8ms/step - loss: 0.0066 - root_mean_squared_error: 0.0815 - val_loss: 0.0125 - val_root_mean_squared_error: 0.1118
Epoch 14/100
7/7 [==============================] - 0s 8ms/step - loss: 0.0059 - root_mean_squared_error: 0.0768 - val_loss: 0.0116 - val_root_mean_squared_error: 0.1079
Epoch 15/100
7/7 [==============================] - 0s 8ms/step - loss: 0.0053 - root_mean_squared_error: 0.0726 - val_loss: 0.0098 - val_root_mean_sq

7/7 [==============================] - 0s 8ms/step - loss: 6.0181e-04 - root_mean_squared_error: 0.0245 - val_loss: 0.0016 - val_root_mean_squared_error: 0.0402
Epoch 11/100
7/7 [==============================] - 0s 8ms/step - loss: 5.4734e-04 - root_mean_squared_error: 0.0234 - val_loss: 0.0015 - val_root_mean_squared_error: 0.0392
Epoch 12/100
7/7 [==============================] - 0s 8ms/step - loss: 5.0435e-04 - root_mean_squared_error: 0.0225 - val_loss: 0.0014 - val_root_mean_squared_error: 0.0376
Epoch 13/100
7/7 [==============================] - 0s 8ms/step - loss: 4.6283e-04 - root_mean_squared_error: 0.0215 - val_loss: 0.0013 - val_root_mean_squared_error: 0.0366
Epoch 14/100
7/7 [==============================] - 0s 8ms/step - loss: 4.2954e-04 - root_mean_squared_error: 0.0207 - val_loss: 0.0013 - val_root_mean_squared_error: 0.0358
Epoch 15/100
7/7 [==============================] - 0s 8ms/step - loss: 4.0252e-04 - root_mean_squared_error: 0.0201 - val_loss: 0.0012 - val_r

7/7 [==============================] - 0s 8ms/step - loss: 7.2250e-04 - root_mean_squared_error: 0.0269 - val_loss: 0.0454 - val_root_mean_squared_error: 0.2131
Epoch 22/100
7/7 [==============================] - 0s 8ms/step - loss: 7.1092e-04 - root_mean_squared_error: 0.0267 - val_loss: 0.0448 - val_root_mean_squared_error: 0.2118
Epoch 23/100
7/7 [==============================] - 0s 8ms/step - loss: 7.0102e-04 - root_mean_squared_error: 0.0265 - val_loss: 0.0445 - val_root_mean_squared_error: 0.2109
Epoch 24/100
7/7 [==============================] - 0s 8ms/step - loss: 6.9072e-04 - root_mean_squared_error: 0.0263 - val_loss: 0.0442 - val_root_mean_squared_error: 0.2103
Epoch 25/100
7/7 [==============================] - 0s 8ms/step - loss: 6.8112e-04 - root_mean_squared_error: 0.0261 - val_loss: 0.0438 - val_root_mean_squared_error: 0.2094
Epoch 26/100
7/7 [==============================] - 0s 8ms/step - loss: 6.7201e-04 - root_mean_squared_error: 0.0259 - val_loss: 0.0436 - val_r

7/7 [==============================] - 0s 8ms/step - loss: 4.8671e-04 - root_mean_squared_error: 0.0221 - val_loss: 0.0337 - val_root_mean_squared_error: 0.1837
Epoch 69/100
7/7 [==============================] - 0s 9ms/step - loss: 4.8444e-04 - root_mean_squared_error: 0.0220 - val_loss: 0.0335 - val_root_mean_squared_error: 0.1832
Epoch 70/100
7/7 [==============================] - 0s 8ms/step - loss: 4.8284e-04 - root_mean_squared_error: 0.0220 - val_loss: 0.0333 - val_root_mean_squared_error: 0.1826
Epoch 71/100
7/7 [==============================] - 0s 9ms/step - loss: 4.8100e-04 - root_mean_squared_error: 0.0219 - val_loss: 0.0333 - val_root_mean_squared_error: 0.1824
Epoch 72/100
7/7 [==============================] - 0s 8ms/step - loss: 4.8002e-04 - root_mean_squared_error: 0.0219 - val_loss: 0.0331 - val_root_mean_squared_error: 0.1819
Epoch 73/100
7/7 [==============================] - 0s 8ms/step - loss: 4.7836e-04 - root_mean_squared_error: 0.0219 - val_loss: 0.0331 - val_r

13/13 [==============================] - 0s 7ms/step - loss: 2.6709e-04 - root_mean_squared_error: 0.0163 - val_loss: 8.7651e-04 - val_root_mean_squared_error: 0.0296
Epoch 8/50
13/13 [==============================] - 0s 7ms/step - loss: 2.5657e-04 - root_mean_squared_error: 0.0160 - val_loss: 8.9040e-04 - val_root_mean_squared_error: 0.0298
Epoch 9/50
7/7 [==============================] - 0s 2ms/step - loss: 6.6668e-05 - root_mean_squared_error: 0.0082
Epoch 1/50
13/13 [==============================] - 1s 17ms/step - loss: 0.0021 - root_mean_squared_error: 0.0457 - val_loss: 0.0033 - val_root_mean_squared_error: 0.0576
Epoch 2/50
13/13 [==============================] - 0s 6ms/step - loss: 2.3205e-04 - root_mean_squared_error: 0.0152 - val_loss: 0.0036 - val_root_mean_squared_error: 0.0598
Epoch 3/50
13/13 [==============================] - 0s 6ms/step - loss: 1.7427e-04 - root_mean_squared_error: 0.0132 - val_loss: 0.0021 - val_root_mean_squared_error: 0.0453
Epoch 4/50
13/13 [===

13/13 [==============================] - 0s 7ms/step - loss: 0.0020 - root_mean_squared_error: 0.0448 - val_loss: 0.0035 - val_root_mean_squared_error: 0.0591
Epoch 39/100
13/13 [==============================] - 0s 7ms/step - loss: 0.0020 - root_mean_squared_error: 0.0443 - val_loss: 0.0034 - val_root_mean_squared_error: 0.0584
Epoch 40/100
13/13 [==============================] - 0s 7ms/step - loss: 0.0019 - root_mean_squared_error: 0.0439 - val_loss: 0.0033 - val_root_mean_squared_error: 0.0571
Epoch 41/100
13/13 [==============================] - 0s 7ms/step - loss: 0.0019 - root_mean_squared_error: 0.0434 - val_loss: 0.0032 - val_root_mean_squared_error: 0.0569
Epoch 42/100
13/13 [==============================] - 0s 7ms/step - loss: 0.0018 - root_mean_squared_error: 0.0430 - val_loss: 0.0032 - val_root_mean_squared_error: 0.0562
Epoch 43/100
13/13 [==============================] - 0s 7ms/step - loss: 0.0018 - root_mean_squared_error: 0.0425 - val_loss: 0.0032 - val_root_mean_squ

13/13 [==============================] - 0s 7ms/step - loss: 9.3244e-04 - root_mean_squared_error: 0.0305 - val_loss: 0.0016 - val_root_mean_squared_error: 0.0401
Epoch 86/100
13/13 [==============================] - 0s 7ms/step - loss: 9.2311e-04 - root_mean_squared_error: 0.0304 - val_loss: 0.0016 - val_root_mean_squared_error: 0.0401
Epoch 87/100
13/13 [==============================] - 0s 6ms/step - loss: 9.1359e-04 - root_mean_squared_error: 0.0302 - val_loss: 0.0016 - val_root_mean_squared_error: 0.0397
Epoch 88/100
13/13 [==============================] - 0s 7ms/step - loss: 9.0522e-04 - root_mean_squared_error: 0.0301 - val_loss: 0.0015 - val_root_mean_squared_error: 0.0393
Epoch 89/100
13/13 [==============================] - 0s 7ms/step - loss: 8.9888e-04 - root_mean_squared_error: 0.0300 - val_loss: 0.0015 - val_root_mean_squared_error: 0.0390
Epoch 90/100
13/13 [==============================] - 0s 7ms/step - loss: 8.8964e-04 - root_mean_squared_error: 0.0298 - val_loss: 0.

13/13 [==============================] - 0s 7ms/step - loss: 7.1364e-04 - root_mean_squared_error: 0.0267 - val_loss: 0.0016 - val_root_mean_squared_error: 0.0403
Epoch 38/100
13/13 [==============================] - 0s 7ms/step - loss: 7.0803e-04 - root_mean_squared_error: 0.0266 - val_loss: 0.0016 - val_root_mean_squared_error: 0.0404
Epoch 39/100
13/13 [==============================] - 0s 7ms/step - loss: 7.1685e-04 - root_mean_squared_error: 0.0268 - val_loss: 0.0016 - val_root_mean_squared_error: 0.0406
Epoch 40/100
13/13 [==============================] - 0s 6ms/step - loss: 6.9343e-04 - root_mean_squared_error: 0.0263 - val_loss: 0.0016 - val_root_mean_squared_error: 0.0402
Epoch 41/100
13/13 [==============================] - 0s 6ms/step - loss: 6.8740e-04 - root_mean_squared_error: 0.0262 - val_loss: 0.0016 - val_root_mean_squared_error: 0.0403
Epoch 42/100
13/13 [==============================] - 0s 7ms/step - loss: 6.7964e-04 - root_mean_squared_error: 0.0261 - val_loss: 0.

13/13 [==============================] - 0s 7ms/step - loss: 2.3363e-04 - root_mean_squared_error: 0.0153 - val_loss: 0.0265 - val_root_mean_squared_error: 0.1629
Epoch 31/100
13/13 [==============================] - 0s 7ms/step - loss: 2.3271e-04 - root_mean_squared_error: 0.0153 - val_loss: 0.0263 - val_root_mean_squared_error: 0.1623
Epoch 32/100
13/13 [==============================] - 0s 7ms/step - loss: 2.3217e-04 - root_mean_squared_error: 0.0152 - val_loss: 0.0261 - val_root_mean_squared_error: 0.1616
Epoch 33/100
13/13 [==============================] - 0s 8ms/step - loss: 2.3142e-04 - root_mean_squared_error: 0.0152 - val_loss: 0.0258 - val_root_mean_squared_error: 0.1607
Epoch 34/100
13/13 [==============================] - 0s 7ms/step - loss: 2.3043e-04 - root_mean_squared_error: 0.0152 - val_loss: 0.0257 - val_root_mean_squared_error: 0.1603
Epoch 35/100
13/13 [==============================] - 0s 8ms/step - loss: 2.2977e-04 - root_mean_squared_error: 0.0152 - val_loss: 0.

13/13 [==============================] - 0s 7ms/step - loss: 2.0934e-04 - root_mean_squared_error: 0.0145 - val_loss: 0.0217 - val_root_mean_squared_error: 0.1475
Epoch 77/100
13/13 [==============================] - 0s 6ms/step - loss: 2.0868e-04 - root_mean_squared_error: 0.0144 - val_loss: 0.0217 - val_root_mean_squared_error: 0.1473
Epoch 78/100
13/13 [==============================] - 0s 7ms/step - loss: 2.0853e-04 - root_mean_squared_error: 0.0144 - val_loss: 0.0216 - val_root_mean_squared_error: 0.1471
Epoch 79/100
13/13 [==============================] - 0s 7ms/step - loss: 2.0802e-04 - root_mean_squared_error: 0.0144 - val_loss: 0.0215 - val_root_mean_squared_error: 0.1468
Epoch 80/100
13/13 [==============================] - 0s 7ms/step - loss: 2.0745e-04 - root_mean_squared_error: 0.0144 - val_loss: 0.0214 - val_root_mean_squared_error: 0.1465
Epoch 81/100
13/13 [==============================] - 0s 7ms/step - loss: 2.0709e-04 - root_mean_squared_error: 0.0144 - val_loss: 0.

7/7 [==============================] - 1s 55ms/step - loss: 0.0054 - root_mean_squared_error: 0.0733 - val_loss: 0.0891 - val_root_mean_squared_error: 0.2984
Epoch 2/50
7/7 [==============================] - 0s 7ms/step - loss: 0.0023 - root_mean_squared_error: 0.0482 - val_loss: 0.0697 - val_root_mean_squared_error: 0.2640
Epoch 3/50
7/7 [==============================] - 0s 7ms/step - loss: 0.0016 - root_mean_squared_error: 0.0405 - val_loss: 0.0482 - val_root_mean_squared_error: 0.2196
Epoch 4/50
7/7 [==============================] - 0s 7ms/step - loss: 0.0011 - root_mean_squared_error: 0.0331 - val_loss: 0.0337 - val_root_mean_squared_error: 0.1836
Epoch 5/50
7/7 [==============================] - 0s 7ms/step - loss: 8.4539e-04 - root_mean_squared_error: 0.0291 - val_loss: 0.0253 - val_root_mean_squared_error: 0.1591
Epoch 6/50
7/7 [==============================] - 0s 8ms/step - loss: 6.1251e-04 - root_mean_squared_error: 0.0247 - val_loss: 0.0214 - val_root_mean_squared_error: 0

In [14]:
sl_LSTM_model = KerasRegressor(single_layer_LSTM_model) 
params = dict(optimizer=['sgd', 'adam'], epochs=[50, 100], batch_size=[16, 32], activation=['relu','tanh'],
             learning_rate=[0.01, 0.001, 0.0005])
random_search = RandomizedSearchCV(sl_LSTM_model, params, cv=3)
sl_LSTM_results = random_search.fit(x_train, y_train, validation_data=(x_valid, y_valid),
                                          callbacks=[EarlyStopping(patience=3)])
print(f"Best: {sl_LSTM_results.best_score_} using {sl_LSTM_results.best_params_}")

<ipython-input-14-83ce690538f2>:1: DeprecationWarning: KerasRegressor is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  sl_LSTM_model = KerasRegressor(single_layer_LSTM_model)


Epoch 1/50
7/7 [==============================] - 2s 102ms/step - loss: 0.0258 - root_mean_squared_error: 0.1606 - val_loss: 0.0186 - val_root_mean_squared_error: 0.1363
Epoch 2/50
7/7 [==============================] - 0s 18ms/step - loss: 0.0038 - root_mean_squared_error: 0.0617 - val_loss: 6.2848e-04 - val_root_mean_squared_error: 0.0251
Epoch 3/50
7/7 [==============================] - 0s 18ms/step - loss: 0.0036 - root_mean_squared_error: 0.0604 - val_loss: 0.0028 - val_root_mean_squared_error: 0.0527
Epoch 4/50
7/7 [==============================] - 0s 18ms/step - loss: 0.0020 - root_mean_squared_error: 0.0445 - val_loss: 0.0061 - val_root_mean_squared_error: 0.0784
Epoch 5/50
4/4 [==============================] - 0s 5ms/step - loss: 0.0032 - root_mean_squared_error: 0.0566
Epoch 1/50
7/7 [==============================] - 2s 68ms/step - loss: 0.0239 - root_mean_squared_error: 0.1546 - val_loss: 0.0246 - val_root_mean_squared_error: 0.1570
Epoch 2/50
7/7 [=======================

Epoch 29/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0052 - root_mean_squared_error: 0.0720 - val_loss: 0.0133 - val_root_mean_squared_error: 0.1153
Epoch 30/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0051 - root_mean_squared_error: 0.0713 - val_loss: 0.0130 - val_root_mean_squared_error: 0.1138
Epoch 31/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0050 - root_mean_squared_error: 0.0707 - val_loss: 0.0126 - val_root_mean_squared_error: 0.1123
Epoch 32/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0049 - root_mean_squared_error: 0.0701 - val_loss: 0.0123 - val_root_mean_squared_error: 0.1109
Epoch 33/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0048 - root_mean_squared_error: 0.0696 - val_loss: 0.0120 - val_root_mean_squared_error: 0.1096
Epoch 34/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0048 - root_mean_squared_error: 0.0691 - val_loss: 0.0118 

13/13 [==============================] - 0s 14ms/step - loss: 0.0041 - root_mean_squared_error: 0.0639 - val_loss: 0.0076 - val_root_mean_squared_error: 0.0869
Epoch 77/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0041 - root_mean_squared_error: 0.0639 - val_loss: 0.0075 - val_root_mean_squared_error: 0.0868
Epoch 78/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0041 - root_mean_squared_error: 0.0639 - val_loss: 0.0075 - val_root_mean_squared_error: 0.0866
Epoch 79/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0041 - root_mean_squared_error: 0.0639 - val_loss: 0.0075 - val_root_mean_squared_error: 0.0865
Epoch 80/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0041 - root_mean_squared_error: 0.0638 - val_loss: 0.0075 - val_root_mean_squared_error: 0.0864
Epoch 81/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0041 - root_mean_squared_error: 0.0638 - val_loss: 0.0074 - val_root_me

Epoch 23/100
13/13 [==============================] - 0s 16ms/step - loss: 0.0113 - root_mean_squared_error: 0.1062 - val_loss: 0.0273 - val_root_mean_squared_error: 0.1653
Epoch 24/100
13/13 [==============================] - 0s 16ms/step - loss: 0.0111 - root_mean_squared_error: 0.1053 - val_loss: 0.0267 - val_root_mean_squared_error: 0.1634
Epoch 25/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0109 - root_mean_squared_error: 0.1045 - val_loss: 0.0261 - val_root_mean_squared_error: 0.1616
Epoch 26/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0108 - root_mean_squared_error: 0.1038 - val_loss: 0.0256 - val_root_mean_squared_error: 0.1599
Epoch 27/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0106 - root_mean_squared_error: 0.1031 - val_loss: 0.0250 - val_root_mean_squared_error: 0.1582
Epoch 28/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0105 - root_mean_squared_error: 0.1025 - val_loss: 0.0245 

13/13 [==============================] - 0s 14ms/step - loss: 0.0087 - root_mean_squared_error: 0.0935 - val_loss: 0.0153 - val_root_mean_squared_error: 0.1239
Epoch 71/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0087 - root_mean_squared_error: 0.0934 - val_loss: 0.0153 - val_root_mean_squared_error: 0.1236
Epoch 72/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0087 - root_mean_squared_error: 0.0933 - val_loss: 0.0152 - val_root_mean_squared_error: 0.1233
Epoch 73/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0087 - root_mean_squared_error: 0.0933 - val_loss: 0.0151 - val_root_mean_squared_error: 0.1230
Epoch 74/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0087 - root_mean_squared_error: 0.0932 - val_loss: 0.0151 - val_root_mean_squared_error: 0.1228
Epoch 75/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0087 - root_mean_squared_error: 0.0932 - val_loss: 0.0150 - val_root_me

Epoch 17/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0025 - root_mean_squared_error: 0.0505 - val_loss: 0.0693 - val_root_mean_squared_error: 0.2633
Epoch 18/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0025 - root_mean_squared_error: 0.0498 - val_loss: 0.0686 - val_root_mean_squared_error: 0.2620
Epoch 19/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0024 - root_mean_squared_error: 0.0492 - val_loss: 0.0680 - val_root_mean_squared_error: 0.2608
Epoch 20/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0024 - root_mean_squared_error: 0.0486 - val_loss: 0.0674 - val_root_mean_squared_error: 0.2596
Epoch 21/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0023 - root_mean_squared_error: 0.0482 - val_loss: 0.0668 - val_root_mean_squared_error: 0.2584
Epoch 22/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0023 - root_mean_squared_error: 0.0477 - val_loss: 0.0662 

13/13 [==============================] - 0s 14ms/step - loss: 0.0019 - root_mean_squared_error: 0.0433 - val_loss: 0.0573 - val_root_mean_squared_error: 0.2393
Epoch 65/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0019 - root_mean_squared_error: 0.0432 - val_loss: 0.0572 - val_root_mean_squared_error: 0.2392
Epoch 66/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0019 - root_mean_squared_error: 0.0432 - val_loss: 0.0572 - val_root_mean_squared_error: 0.2391
Epoch 67/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0019 - root_mean_squared_error: 0.0432 - val_loss: 0.0571 - val_root_mean_squared_error: 0.2390
Epoch 68/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0019 - root_mean_squared_error: 0.0432 - val_loss: 0.0571 - val_root_mean_squared_error: 0.2389
Epoch 69/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0019 - root_mean_squared_error: 0.0432 - val_loss: 0.0570 - val_root_me

7/7 [==============================] - 0s 17ms/step - loss: 0.0056 - root_mean_squared_error: 0.0749 - val_loss: 0.0097 - val_root_mean_squared_error: 0.0984
Epoch 12/50
7/7 [==============================] - 0s 16ms/step - loss: 0.0056 - root_mean_squared_error: 0.0747 - val_loss: 0.0099 - val_root_mean_squared_error: 0.0997
Epoch 13/50
7/7 [==============================] - 0s 16ms/step - loss: 0.0056 - root_mean_squared_error: 0.0746 - val_loss: 0.0095 - val_root_mean_squared_error: 0.0977
Epoch 14/50
7/7 [==============================] - 0s 17ms/step - loss: 0.0055 - root_mean_squared_error: 0.0743 - val_loss: 0.0096 - val_root_mean_squared_error: 0.0979
Epoch 15/50
7/7 [==============================] - 0s 18ms/step - loss: 0.0055 - root_mean_squared_error: 0.0742 - val_loss: 0.0096 - val_root_mean_squared_error: 0.0981
Epoch 16/50
4/4 [==============================] - 0s 4ms/step - loss: 0.0150 - root_mean_squared_error: 0.1223
Epoch 1/50
7/7 [==============================] - 

7/7 [==============================] - 0s 14ms/step - loss: 0.0015 - root_mean_squared_error: 0.0384 - val_loss: 0.0447 - val_root_mean_squared_error: 0.2115
Epoch 30/50
7/7 [==============================] - 0s 15ms/step - loss: 0.0015 - root_mean_squared_error: 0.0384 - val_loss: 0.0446 - val_root_mean_squared_error: 0.2113
Epoch 31/50
7/7 [==============================] - 0s 15ms/step - loss: 0.0015 - root_mean_squared_error: 0.0384 - val_loss: 0.0446 - val_root_mean_squared_error: 0.2111
Epoch 32/50
7/7 [==============================] - 0s 14ms/step - loss: 0.0015 - root_mean_squared_error: 0.0383 - val_loss: 0.0444 - val_root_mean_squared_error: 0.2106
Epoch 33/50
7/7 [==============================] - 0s 14ms/step - loss: 0.0015 - root_mean_squared_error: 0.0383 - val_loss: 0.0447 - val_root_mean_squared_error: 0.2115
Epoch 34/50
7/7 [==============================] - 0s 15ms/step - loss: 0.0015 - root_mean_squared_error: 0.0383 - val_loss: 0.0449 - val_root_mean_squared_error:

13/13 [==============================] - 0s 13ms/step - loss: 7.1084e-05 - root_mean_squared_error: 0.0084 - val_loss: 0.0035 - val_root_mean_squared_error: 0.0594
Epoch 5/50
13/13 [==============================] - 0s 13ms/step - loss: 3.7904e-05 - root_mean_squared_error: 0.0062 - val_loss: 6.8637e-04 - val_root_mean_squared_error: 0.0262
Epoch 6/50
13/13 [==============================] - 0s 13ms/step - loss: 3.0284e-05 - root_mean_squared_error: 0.0055 - val_loss: 0.0011 - val_root_mean_squared_error: 0.0326
Epoch 7/50
13/13 [==============================] - 0s 14ms/step - loss: 2.8076e-05 - root_mean_squared_error: 0.0053 - val_loss: 0.0011 - val_root_mean_squared_error: 0.0330
Epoch 8/50
7/7 [==============================] - 0s 3ms/step - loss: 0.0023 - root_mean_squared_error: 0.0479
Epoch 1/100
7/7 [==============================] - 2s 74ms/step - loss: 0.0166 - root_mean_squared_error: 0.1288 - val_loss: 0.0088 - val_root_mean_squared_error: 0.0939
Epoch 2/100
7/7 [=========

7/7 [==============================] - 0s 18ms/step - loss: 7.0163e-05 - root_mean_squared_error: 0.0084 - val_loss: 2.9029e-04 - val_root_mean_squared_error: 0.0170
Epoch 6/100
7/7 [==============================] - 0s 16ms/step - loss: 6.2565e-05 - root_mean_squared_error: 0.0079 - val_loss: 2.4027e-04 - val_root_mean_squared_error: 0.0155
Epoch 7/100
7/7 [==============================] - 0s 16ms/step - loss: 3.7707e-05 - root_mean_squared_error: 0.0061 - val_loss: 2.2361e-04 - val_root_mean_squared_error: 0.0150
Epoch 8/100
7/7 [==============================] - 0s 15ms/step - loss: 3.5960e-05 - root_mean_squared_error: 0.0060 - val_loss: 2.1476e-04 - val_root_mean_squared_error: 0.0147
Epoch 9/100
7/7 [==============================] - 0s 16ms/step - loss: 1.7485e-05 - root_mean_squared_error: 0.0042 - val_loss: 3.4757e-04 - val_root_mean_squared_error: 0.0186
Epoch 10/100
7/7 [==============================] - 0s 18ms/step - loss: 2.6585e-05 - root_mean_squared_error: 0.0052 - va

Epoch 41/100
13/13 [==============================] - 0s 11ms/step - loss: 0.0036 - root_mean_squared_error: 0.0600 - val_loss: 0.0071 - val_root_mean_squared_error: 0.0845
Epoch 42/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0036 - root_mean_squared_error: 0.0600 - val_loss: 0.0071 - val_root_mean_squared_error: 0.0840
Epoch 43/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0036 - root_mean_squared_error: 0.0599 - val_loss: 0.0070 - val_root_mean_squared_error: 0.0836
Epoch 44/100
13/13 [==============================] - 0s 11ms/step - loss: 0.0036 - root_mean_squared_error: 0.0599 - val_loss: 0.0069 - val_root_mean_squared_error: 0.0832
Epoch 45/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0036 - root_mean_squared_error: 0.0598 - val_loss: 0.0069 - val_root_mean_squared_error: 0.0829
Epoch 46/100
13/13 [==============================] - 0s 11ms/step - loss: 0.0036 - root_mean_squared_error: 0.0597 - val_loss: 0.0068 

Epoch 6/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0178 - root_mean_squared_error: 0.1334 - val_loss: 0.0439 - val_root_mean_squared_error: 0.2095
Epoch 7/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0167 - root_mean_squared_error: 0.1293 - val_loss: 0.0412 - val_root_mean_squared_error: 0.2030
Epoch 8/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0158 - root_mean_squared_error: 0.1259 - val_loss: 0.0389 - val_root_mean_squared_error: 0.1971
Epoch 9/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0151 - root_mean_squared_error: 0.1229 - val_loss: 0.0368 - val_root_mean_squared_error: 0.1918
Epoch 10/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0145 - root_mean_squared_error: 0.1203 - val_loss: 0.0349 - val_root_mean_squared_error: 0.1868
Epoch 11/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0139 - root_mean_squared_error: 0.1180 - val_loss: 0.0332 - va

13/13 [==============================] - 0s 13ms/step - loss: 0.0102 - root_mean_squared_error: 0.1010 - val_loss: 0.0164 - val_root_mean_squared_error: 0.1280
Epoch 54/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0102 - root_mean_squared_error: 0.1009 - val_loss: 0.0163 - val_root_mean_squared_error: 0.1278
Epoch 55/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0102 - root_mean_squared_error: 0.1008 - val_loss: 0.0163 - val_root_mean_squared_error: 0.1275
Epoch 56/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0101 - root_mean_squared_error: 0.1007 - val_loss: 0.0162 - val_root_mean_squared_error: 0.1274
Epoch 57/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0101 - root_mean_squared_error: 0.1006 - val_loss: 0.0161 - val_root_mean_squared_error: 0.1270
Epoch 58/100
13/13 [==============================] - 0s 12ms/step - loss: 0.0101 - root_mean_squared_error: 0.1005 - val_loss: 0.0161 - val_root_me

7/7 [==============================] - 0s 4ms/step - loss: 0.0023 - root_mean_squared_error: 0.0481
Epoch 1/100
13/13 [==============================] - 2s 41ms/step - loss: 0.0022 - root_mean_squared_error: 0.0471 - val_loss: 0.0356 - val_root_mean_squared_error: 0.1888
Epoch 2/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0020 - root_mean_squared_error: 0.0451 - val_loss: 0.0347 - val_root_mean_squared_error: 0.1863
Epoch 3/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0019 - root_mean_squared_error: 0.0432 - val_loss: 0.0338 - val_root_mean_squared_error: 0.1840
Epoch 4/100
13/13 [==============================] - 0s 13ms/step - loss: 0.0017 - root_mean_squared_error: 0.0415 - val_loss: 0.0330 - val_root_mean_squared_error: 0.1817
Epoch 5/100
13/13 [==============================] - 0s 14ms/step - loss: 0.0016 - root_mean_squared_error: 0.0400 - val_loss: 0.0323 - val_root_mean_squared_error: 0.1797
Epoch 6/100
13/13 [=====================

13/13 [==============================] - 0s 13ms/step - loss: 7.9117e-04 - root_mean_squared_error: 0.0281 - val_loss: 0.0235 - val_root_mean_squared_error: 0.1532
Epoch 47/100
13/13 [==============================] - 0s 13ms/step - loss: 7.9079e-04 - root_mean_squared_error: 0.0281 - val_loss: 0.0234 - val_root_mean_squared_error: 0.1530
Epoch 48/100
13/13 [==============================] - 0s 12ms/step - loss: 7.9031e-04 - root_mean_squared_error: 0.0281 - val_loss: 0.0234 - val_root_mean_squared_error: 0.1530
Epoch 49/100
13/13 [==============================] - 0s 13ms/step - loss: 7.9012e-04 - root_mean_squared_error: 0.0281 - val_loss: 0.0234 - val_root_mean_squared_error: 0.1529
Epoch 50/100
13/13 [==============================] - 0s 12ms/step - loss: 7.8965e-04 - root_mean_squared_error: 0.0281 - val_loss: 0.0233 - val_root_mean_squared_error: 0.1528
Epoch 51/100
13/13 [==============================] - 0s 12ms/step - loss: 7.8940e-04 - root_mean_squared_error: 0.0281 - val_lo

13/13 [==============================] - 0s 13ms/step - loss: 7.8175e-04 - root_mean_squared_error: 0.0280 - val_loss: 0.0228 - val_root_mean_squared_error: 0.1511
Epoch 93/100
13/13 [==============================] - 0s 13ms/step - loss: 7.8134e-04 - root_mean_squared_error: 0.0280 - val_loss: 0.0228 - val_root_mean_squared_error: 0.1511
Epoch 94/100
13/13 [==============================] - 0s 13ms/step - loss: 7.8123e-04 - root_mean_squared_error: 0.0280 - val_loss: 0.0228 - val_root_mean_squared_error: 0.1510
Epoch 95/100
13/13 [==============================] - 0s 13ms/step - loss: 7.8101e-04 - root_mean_squared_error: 0.0279 - val_loss: 0.0228 - val_root_mean_squared_error: 0.1510
Epoch 96/100
13/13 [==============================] - 0s 13ms/step - loss: 7.8104e-04 - root_mean_squared_error: 0.0279 - val_loss: 0.0228 - val_root_mean_squared_error: 0.1510
Epoch 97/100
13/13 [==============================] - 0s 13ms/step - loss: 7.8076e-04 - root_mean_squared_error: 0.0279 - val_lo

7/7 [==============================] - 0s 15ms/step - loss: 0.0057 - root_mean_squared_error: 0.0757 - val_loss: 0.0155 - val_root_mean_squared_error: 0.1246
Epoch 42/50
7/7 [==============================] - 0s 15ms/step - loss: 0.0056 - root_mean_squared_error: 0.0750 - val_loss: 0.0153 - val_root_mean_squared_error: 0.1236
Epoch 43/50
7/7 [==============================] - 0s 15ms/step - loss: 0.0055 - root_mean_squared_error: 0.0744 - val_loss: 0.0150 - val_root_mean_squared_error: 0.1226
Epoch 44/50
7/7 [==============================] - 0s 15ms/step - loss: 0.0054 - root_mean_squared_error: 0.0738 - val_loss: 0.0148 - val_root_mean_squared_error: 0.1217
Epoch 45/50
7/7 [==============================] - 0s 15ms/step - loss: 0.0054 - root_mean_squared_error: 0.0732 - val_loss: 0.0146 - val_root_mean_squared_error: 0.1206
Epoch 46/50
7/7 [==============================] - 0s 16ms/step - loss: 0.0053 - root_mean_squared_error: 0.0726 - val_loss: 0.0143 - val_root_mean_squared_error:

Epoch 39/50
7/7 [==============================] - 0s 16ms/step - loss: 0.0126 - root_mean_squared_error: 0.1124 - val_loss: 0.0306 - val_root_mean_squared_error: 0.1750
Epoch 40/50
7/7 [==============================] - 0s 17ms/step - loss: 0.0125 - root_mean_squared_error: 0.1118 - val_loss: 0.0302 - val_root_mean_squared_error: 0.1737
Epoch 41/50
7/7 [==============================] - 0s 17ms/step - loss: 0.0124 - root_mean_squared_error: 0.1113 - val_loss: 0.0298 - val_root_mean_squared_error: 0.1726
Epoch 42/50
7/7 [==============================] - 0s 19ms/step - loss: 0.0123 - root_mean_squared_error: 0.1108 - val_loss: 0.0294 - val_root_mean_squared_error: 0.1715
Epoch 43/50
7/7 [==============================] - 0s 18ms/step - loss: 0.0122 - root_mean_squared_error: 0.1103 - val_loss: 0.0290 - val_root_mean_squared_error: 0.1704
Epoch 44/50
7/7 [==============================] - 0s 18ms/step - loss: 0.0121 - root_mean_squared_error: 0.1099 - val_loss: 0.0287 - val_root_mean_sq

Epoch 37/50
7/7 [==============================] - 0s 16ms/step - loss: 0.0032 - root_mean_squared_error: 0.0569 - val_loss: 0.0896 - val_root_mean_squared_error: 0.2993
Epoch 38/50
7/7 [==============================] - 0s 15ms/step - loss: 0.0032 - root_mean_squared_error: 0.0566 - val_loss: 0.0892 - val_root_mean_squared_error: 0.2987
Epoch 39/50
7/7 [==============================] - 0s 16ms/step - loss: 0.0032 - root_mean_squared_error: 0.0563 - val_loss: 0.0888 - val_root_mean_squared_error: 0.2980
Epoch 40/50
7/7 [==============================] - 0s 15ms/step - loss: 0.0031 - root_mean_squared_error: 0.0560 - val_loss: 0.0885 - val_root_mean_squared_error: 0.2974
Epoch 41/50
7/7 [==============================] - 0s 16ms/step - loss: 0.0031 - root_mean_squared_error: 0.0557 - val_loss: 0.0881 - val_root_mean_squared_error: 0.2968
Epoch 42/50
7/7 [==============================] - 0s 15ms/step - loss: 0.0031 - root_mean_squared_error: 0.0554 - val_loss: 0.0877 - val_root_mean_sq

In [15]:
ml_LSTM_model = KerasRegressor(multiple_layer_LSTM_model) 
params = dict(optimizer=['sgd', 'adam'], epochs=[50, 100], batch_size=[16, 32], activation=['relu','tanh'],
             learning_rate=[0.01, 0.001, 0.0005])
random_search = RandomizedSearchCV(ml_LSTM_model, params, cv=3)
ml_LSTM_results = random_search.fit(x_train, y_train, validation_data=(x_valid, y_valid),
                                          callbacks=[EarlyStopping(patience=3)])
print(f"Best: {ml_LSTM_results.best_score_} using {ml_LSTM_results.best_params_}")

<ipython-input-15-369a7f03016f>:1: DeprecationWarning: KerasRegressor is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  ml_LSTM_model = KerasRegressor(multiple_layer_LSTM_model)


Epoch 1/50
7/7 [==============================] - 3s 112ms/step - loss: 0.0102 - root_mean_squared_error: 0.1012 - val_loss: 0.0025 - val_root_mean_squared_error: 0.0499
Epoch 2/50
7/7 [==============================] - 0s 26ms/step - loss: 0.0034 - root_mean_squared_error: 0.0587 - val_loss: 0.0061 - val_root_mean_squared_error: 0.0780
Epoch 3/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0023 - root_mean_squared_error: 0.0483 - val_loss: 0.0031 - val_root_mean_squared_error: 0.0560
Epoch 4/50
7/7 [==============================] - 0s 28ms/step - loss: 0.0013 - root_mean_squared_error: 0.0358 - val_loss: 2.3876e-04 - val_root_mean_squared_error: 0.0155
Epoch 5/50
7/7 [==============================] - 0s 28ms/step - loss: 9.8085e-04 - root_mean_squared_error: 0.0313 - val_loss: 0.0018 - val_root_mean_squared_error: 0.0424
Epoch 6/50
7/7 [==============================] - 0s 28ms/step - loss: 7.0526e-04 - root_mean_squared_error: 0.0266 - val_loss: 5.2409e-04 - val_r

Epoch 2/100
7/7 [==============================] - 0s 27ms/step - loss: 0.0022 - root_mean_squared_error: 0.0473 - val_loss: 0.0549 - val_root_mean_squared_error: 0.2342
Epoch 3/100
7/7 [==============================] - 0s 30ms/step - loss: 0.0018 - root_mean_squared_error: 0.0420 - val_loss: 0.0517 - val_root_mean_squared_error: 0.2273
Epoch 4/100
7/7 [==============================] - 0s 31ms/step - loss: 0.0016 - root_mean_squared_error: 0.0403 - val_loss: 0.0496 - val_root_mean_squared_error: 0.2227
Epoch 5/100
7/7 [==============================] - 0s 30ms/step - loss: 0.0016 - root_mean_squared_error: 0.0397 - val_loss: 0.0484 - val_root_mean_squared_error: 0.2200
Epoch 6/100
7/7 [==============================] - 0s 31ms/step - loss: 0.0016 - root_mean_squared_error: 0.0395 - val_loss: 0.0482 - val_root_mean_squared_error: 0.2196
Epoch 7/100
7/7 [==============================] - 0s 31ms/step - loss: 0.0016 - root_mean_squared_error: 0.0395 - val_loss: 0.0472 - val_root_mean_sq

Epoch 9/100
7/7 [==============================] - 0s 30ms/step - loss: 0.0120 - root_mean_squared_error: 0.1094 - val_loss: 0.0300 - val_root_mean_squared_error: 0.1733
Epoch 10/100
7/7 [==============================] - 0s 29ms/step - loss: 0.0112 - root_mean_squared_error: 0.1057 - val_loss: 0.0285 - val_root_mean_squared_error: 0.1687
Epoch 11/100
7/7 [==============================] - 0s 29ms/step - loss: 0.0105 - root_mean_squared_error: 0.1025 - val_loss: 0.0270 - val_root_mean_squared_error: 0.1644
Epoch 12/100
7/7 [==============================] - 0s 29ms/step - loss: 0.0099 - root_mean_squared_error: 0.0994 - val_loss: 0.0257 - val_root_mean_squared_error: 0.1602
Epoch 13/100
7/7 [==============================] - 0s 29ms/step - loss: 0.0093 - root_mean_squared_error: 0.0966 - val_loss: 0.0244 - val_root_mean_squared_error: 0.1561
Epoch 14/100
7/7 [==============================] - 0s 29ms/step - loss: 0.0088 - root_mean_squared_error: 0.0939 - val_loss: 0.0232 - val_root_me

Epoch 57/100
7/7 [==============================] - 0s 30ms/step - loss: 0.0050 - root_mean_squared_error: 0.0704 - val_loss: 0.0094 - val_root_mean_squared_error: 0.0972
Epoch 58/100
7/7 [==============================] - 0s 30ms/step - loss: 0.0049 - root_mean_squared_error: 0.0703 - val_loss: 0.0094 - val_root_mean_squared_error: 0.0971
Epoch 59/100
7/7 [==============================] - 0s 30ms/step - loss: 0.0049 - root_mean_squared_error: 0.0703 - val_loss: 0.0094 - val_root_mean_squared_error: 0.0968
Epoch 60/100
7/7 [==============================] - 0s 30ms/step - loss: 0.0049 - root_mean_squared_error: 0.0703 - val_loss: 0.0093 - val_root_mean_squared_error: 0.0962
Epoch 61/100
7/7 [==============================] - 0s 29ms/step - loss: 0.0049 - root_mean_squared_error: 0.0702 - val_loss: 0.0093 - val_root_mean_squared_error: 0.0962
Epoch 62/100
7/7 [==============================] - 0s 29ms/step - loss: 0.0049 - root_mean_squared_error: 0.0702 - val_loss: 0.0092 - val_root_m

7/7 [==============================] - 0s 32ms/step - loss: 0.0099 - root_mean_squared_error: 0.0997 - val_loss: 0.0212 - val_root_mean_squared_error: 0.1456
Epoch 35/100
7/7 [==============================] - 0s 31ms/step - loss: 0.0099 - root_mean_squared_error: 0.0993 - val_loss: 0.0209 - val_root_mean_squared_error: 0.1445
Epoch 36/100
7/7 [==============================] - 0s 31ms/step - loss: 0.0098 - root_mean_squared_error: 0.0990 - val_loss: 0.0205 - val_root_mean_squared_error: 0.1433
Epoch 37/100
7/7 [==============================] - 0s 32ms/step - loss: 0.0098 - root_mean_squared_error: 0.0988 - val_loss: 0.0202 - val_root_mean_squared_error: 0.1423
Epoch 38/100
7/7 [==============================] - 0s 33ms/step - loss: 0.0097 - root_mean_squared_error: 0.0985 - val_loss: 0.0200 - val_root_mean_squared_error: 0.1415
Epoch 39/100
7/7 [==============================] - 0s 35ms/step - loss: 0.0097 - root_mean_squared_error: 0.0984 - val_loss: 0.0198 - val_root_mean_squared_e

Epoch 82/100
7/7 [==============================] - 0s 38ms/step - loss: 0.0091 - root_mean_squared_error: 0.0952 - val_loss: 0.0152 - val_root_mean_squared_error: 0.1232
Epoch 83/100
7/7 [==============================] - 0s 33ms/step - loss: 0.0091 - root_mean_squared_error: 0.0952 - val_loss: 0.0152 - val_root_mean_squared_error: 0.1232
Epoch 84/100
7/7 [==============================] - 0s 31ms/step - loss: 0.0091 - root_mean_squared_error: 0.0952 - val_loss: 0.0151 - val_root_mean_squared_error: 0.1230
Epoch 85/100
7/7 [==============================] - 0s 31ms/step - loss: 0.0090 - root_mean_squared_error: 0.0951 - val_loss: 0.0151 - val_root_mean_squared_error: 0.1228
Epoch 86/100
7/7 [==============================] - 0s 31ms/step - loss: 0.0090 - root_mean_squared_error: 0.0951 - val_loss: 0.0150 - val_root_mean_squared_error: 0.1227
Epoch 87/100
7/7 [==============================] - 0s 31ms/step - loss: 0.0090 - root_mean_squared_error: 0.0950 - val_loss: 0.0150 - val_root_m

7/7 [==============================] - 0s 30ms/step - loss: 0.0019 - root_mean_squared_error: 0.0434 - val_loss: 0.0593 - val_root_mean_squared_error: 0.2435
Epoch 33/100
7/7 [==============================] - 0s 30ms/step - loss: 0.0019 - root_mean_squared_error: 0.0433 - val_loss: 0.0591 - val_root_mean_squared_error: 0.2431
Epoch 34/100
7/7 [==============================] - 0s 29ms/step - loss: 0.0019 - root_mean_squared_error: 0.0433 - val_loss: 0.0588 - val_root_mean_squared_error: 0.2425
Epoch 35/100
7/7 [==============================] - 0s 30ms/step - loss: 0.0019 - root_mean_squared_error: 0.0432 - val_loss: 0.0587 - val_root_mean_squared_error: 0.2422
Epoch 36/100
7/7 [==============================] - 0s 30ms/step - loss: 0.0019 - root_mean_squared_error: 0.0431 - val_loss: 0.0584 - val_root_mean_squared_error: 0.2417
Epoch 37/100
7/7 [==============================] - 0s 30ms/step - loss: 0.0019 - root_mean_squared_error: 0.0430 - val_loss: 0.0582 - val_root_mean_squared_e

Epoch 80/100
7/7 [==============================] - 0s 36ms/step - loss: 0.0018 - root_mean_squared_error: 0.0426 - val_loss: 0.0555 - val_root_mean_squared_error: 0.2356
Epoch 81/100
7/7 [==============================] - 0s 32ms/step - loss: 0.0018 - root_mean_squared_error: 0.0425 - val_loss: 0.0555 - val_root_mean_squared_error: 0.2355
Epoch 82/100
7/7 [==============================] - 0s 31ms/step - loss: 0.0018 - root_mean_squared_error: 0.0425 - val_loss: 0.0554 - val_root_mean_squared_error: 0.2353
Epoch 83/100
7/7 [==============================] - 0s 30ms/step - loss: 0.0018 - root_mean_squared_error: 0.0425 - val_loss: 0.0554 - val_root_mean_squared_error: 0.2353
Epoch 84/100
7/7 [==============================] - 0s 29ms/step - loss: 0.0018 - root_mean_squared_error: 0.0425 - val_loss: 0.0554 - val_root_mean_squared_error: 0.2354
Epoch 85/100
7/7 [==============================] - 0s 29ms/step - loss: 0.0018 - root_mean_squared_error: 0.0425 - val_loss: 0.0554 - val_root_m

7/7 [==============================] - 0s 27ms/step - loss: 0.0073 - root_mean_squared_error: 0.0854 - val_loss: 0.0160 - val_root_mean_squared_error: 0.1267
Epoch 42/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0073 - root_mean_squared_error: 0.0853 - val_loss: 0.0158 - val_root_mean_squared_error: 0.1258
Epoch 43/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0072 - root_mean_squared_error: 0.0851 - val_loss: 0.0156 - val_root_mean_squared_error: 0.1251
Epoch 44/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0072 - root_mean_squared_error: 0.0849 - val_loss: 0.0154 - val_root_mean_squared_error: 0.1242
Epoch 45/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0072 - root_mean_squared_error: 0.0847 - val_loss: 0.0152 - val_root_mean_squared_error: 0.1233
Epoch 46/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0071 - root_mean_squared_error: 0.0845 - val_loss: 0.0150 - val_root_mean_squared_error:

Epoch 39/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0120 - root_mean_squared_error: 0.1097 - val_loss: 0.0213 - val_root_mean_squared_error: 0.1458
Epoch 40/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0120 - root_mean_squared_error: 0.1096 - val_loss: 0.0211 - val_root_mean_squared_error: 0.1452
Epoch 41/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0120 - root_mean_squared_error: 0.1095 - val_loss: 0.0209 - val_root_mean_squared_error: 0.1446
Epoch 42/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0120 - root_mean_squared_error: 0.1094 - val_loss: 0.0208 - val_root_mean_squared_error: 0.1442
Epoch 43/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0119 - root_mean_squared_error: 0.1093 - val_loss: 0.0206 - val_root_mean_squared_error: 0.1434
Epoch 44/50
7/7 [==============================] - 0s 28ms/step - loss: 0.0119 - root_mean_squared_error: 0.1092 - val_loss: 0.0204 - val_root_mean_sq

Epoch 37/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0019 - root_mean_squared_error: 0.0434 - val_loss: 0.0593 - val_root_mean_squared_error: 0.2436
Epoch 38/50
7/7 [==============================] - 0s 28ms/step - loss: 0.0019 - root_mean_squared_error: 0.0433 - val_loss: 0.0591 - val_root_mean_squared_error: 0.2432
Epoch 39/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0019 - root_mean_squared_error: 0.0433 - val_loss: 0.0590 - val_root_mean_squared_error: 0.2429
Epoch 40/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0019 - root_mean_squared_error: 0.0432 - val_loss: 0.0589 - val_root_mean_squared_error: 0.2427
Epoch 41/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0019 - root_mean_squared_error: 0.0432 - val_loss: 0.0588 - val_root_mean_squared_error: 0.2424
Epoch 42/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0019 - root_mean_squared_error: 0.0431 - val_loss: 0.0586 - val_root_mean_sq

Epoch 35/50
7/7 [==============================] - 0s 29ms/step - loss: 0.0106 - root_mean_squared_error: 0.1029 - val_loss: 0.0269 - val_root_mean_squared_error: 0.1639
Epoch 36/50
7/7 [==============================] - 0s 28ms/step - loss: 0.0104 - root_mean_squared_error: 0.1020 - val_loss: 0.0263 - val_root_mean_squared_error: 0.1622
Epoch 37/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0103 - root_mean_squared_error: 0.1013 - val_loss: 0.0258 - val_root_mean_squared_error: 0.1606
Epoch 38/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0101 - root_mean_squared_error: 0.1006 - val_loss: 0.0253 - val_root_mean_squared_error: 0.1591
Epoch 39/50
7/7 [==============================] - 0s 28ms/step - loss: 0.0100 - root_mean_squared_error: 0.0999 - val_loss: 0.0248 - val_root_mean_squared_error: 0.1574
Epoch 40/50
7/7 [==============================] - 0s 27ms/step - loss: 0.0099 - root_mean_squared_error: 0.0993 - val_loss: 0.0243 - val_root_mean_sq

Epoch 33/50
7/7 [==============================] - 0s 28ms/step - loss: 0.0188 - root_mean_squared_error: 0.1371 - val_loss: 0.0437 - val_root_mean_squared_error: 0.2090
Epoch 34/50
7/7 [==============================] - 0s 28ms/step - loss: 0.0186 - root_mean_squared_error: 0.1364 - val_loss: 0.0431 - val_root_mean_squared_error: 0.2075
Epoch 35/50
7/7 [==============================] - 0s 28ms/step - loss: 0.0185 - root_mean_squared_error: 0.1359 - val_loss: 0.0424 - val_root_mean_squared_error: 0.2059
Epoch 36/50
7/7 [==============================] - 0s 28ms/step - loss: 0.0183 - root_mean_squared_error: 0.1352 - val_loss: 0.0418 - val_root_mean_squared_error: 0.2045
Epoch 37/50
7/7 [==============================] - 0s 28ms/step - loss: 0.0181 - root_mean_squared_error: 0.1347 - val_loss: 0.0412 - val_root_mean_squared_error: 0.2030
Epoch 38/50
7/7 [==============================] - 0s 28ms/step - loss: 0.0180 - root_mean_squared_error: 0.1342 - val_loss: 0.0406 - val_root_mean_sq

Epoch 31/50
7/7 [==============================] - 0s 30ms/step - loss: 0.0021 - root_mean_squared_error: 0.0453 - val_loss: 0.0554 - val_root_mean_squared_error: 0.2354
Epoch 32/50
7/7 [==============================] - 0s 30ms/step - loss: 0.0020 - root_mean_squared_error: 0.0450 - val_loss: 0.0551 - val_root_mean_squared_error: 0.2347
Epoch 33/50
7/7 [==============================] - 0s 29ms/step - loss: 0.0020 - root_mean_squared_error: 0.0446 - val_loss: 0.0548 - val_root_mean_squared_error: 0.2341
Epoch 34/50
7/7 [==============================] - 0s 29ms/step - loss: 0.0020 - root_mean_squared_error: 0.0443 - val_loss: 0.0545 - val_root_mean_squared_error: 0.2335
Epoch 35/50
7/7 [==============================] - 0s 29ms/step - loss: 0.0019 - root_mean_squared_error: 0.0440 - val_loss: 0.0542 - val_root_mean_squared_error: 0.2329
Epoch 36/50
7/7 [==============================] - 0s 30ms/step - loss: 0.0019 - root_mean_squared_error: 0.0437 - val_loss: 0.0540 - val_root_mean_sq

13/13 [==============================] - 0s 26ms/step - loss: 0.0019 - root_mean_squared_error: 0.0440 - val_loss: 0.0594 - val_root_mean_squared_error: 0.2437
Epoch 5/50
13/13 [==============================] - 0s 26ms/step - loss: 0.0019 - root_mean_squared_error: 0.0439 - val_loss: 0.0598 - val_root_mean_squared_error: 0.2445
Epoch 6/50
13/13 [==============================] - 0s 24ms/step - loss: 0.0019 - root_mean_squared_error: 0.0438 - val_loss: 0.0602 - val_root_mean_squared_error: 0.2454
Epoch 7/50
7/7 [==============================] - 0s 5ms/step - loss: 0.0461 - root_mean_squared_error: 0.2146
Epoch 1/100
13/13 [==============================] - 3s 71ms/step - loss: 0.0259 - root_mean_squared_error: 0.1608 - val_loss: 0.0553 - val_root_mean_squared_error: 0.2352
Epoch 2/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0238 - root_mean_squared_error: 0.1543 - val_loss: 0.0520 - val_root_mean_squared_error: 0.2280
Epoch 3/100
13/13 [========================

Epoch 44/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0056 - root_mean_squared_error: 0.0749 - val_loss: 0.0122 - val_root_mean_squared_error: 0.1103
Epoch 45/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0056 - root_mean_squared_error: 0.0748 - val_loss: 0.0120 - val_root_mean_squared_error: 0.1095
Epoch 46/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0056 - root_mean_squared_error: 0.0746 - val_loss: 0.0119 - val_root_mean_squared_error: 0.1089
Epoch 47/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0056 - root_mean_squared_error: 0.0745 - val_loss: 0.0117 - val_root_mean_squared_error: 0.1082
Epoch 48/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0055 - root_mean_squared_error: 0.0744 - val_loss: 0.0116 - val_root_mean_squared_error: 0.1076
Epoch 49/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0055 - root_mean_squared_error: 0.0743 - val_loss: 0.0114 

13/13 [==============================] - 0s 24ms/step - loss: 0.0053 - root_mean_squared_error: 0.0729 - val_loss: 0.0094 - val_root_mean_squared_error: 0.0971
Epoch 92/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0053 - root_mean_squared_error: 0.0728 - val_loss: 0.0094 - val_root_mean_squared_error: 0.0970
Epoch 93/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0053 - root_mean_squared_error: 0.0728 - val_loss: 0.0094 - val_root_mean_squared_error: 0.0970
Epoch 94/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0053 - root_mean_squared_error: 0.0728 - val_loss: 0.0094 - val_root_mean_squared_error: 0.0969
Epoch 95/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0053 - root_mean_squared_error: 0.0728 - val_loss: 0.0094 - val_root_mean_squared_error: 0.0969
Epoch 96/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0053 - root_mean_squared_error: 0.0728 - val_loss: 0.0094 - val_root_me

Epoch 38/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0097 - root_mean_squared_error: 0.0986 - val_loss: 0.0180 - val_root_mean_squared_error: 0.1342
Epoch 39/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0097 - root_mean_squared_error: 0.0985 - val_loss: 0.0178 - val_root_mean_squared_error: 0.1334
Epoch 40/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0097 - root_mean_squared_error: 0.0984 - val_loss: 0.0176 - val_root_mean_squared_error: 0.1328
Epoch 41/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0097 - root_mean_squared_error: 0.0982 - val_loss: 0.0175 - val_root_mean_squared_error: 0.1321
Epoch 42/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0096 - root_mean_squared_error: 0.0981 - val_loss: 0.0173 - val_root_mean_squared_error: 0.1314
Epoch 43/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0096 - root_mean_squared_error: 0.0980 - val_loss: 0.0171 

13/13 [==============================] - 0s 24ms/step - loss: 0.0090 - root_mean_squared_error: 0.0949 - val_loss: 0.0144 - val_root_mean_squared_error: 0.1199
Epoch 86/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0090 - root_mean_squared_error: 0.0948 - val_loss: 0.0143 - val_root_mean_squared_error: 0.1196
Epoch 87/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0090 - root_mean_squared_error: 0.0948 - val_loss: 0.0143 - val_root_mean_squared_error: 0.1196
Epoch 88/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0090 - root_mean_squared_error: 0.0947 - val_loss: 0.0143 - val_root_mean_squared_error: 0.1194
Epoch 89/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0090 - root_mean_squared_error: 0.0947 - val_loss: 0.0142 - val_root_mean_squared_error: 0.1193
Epoch 90/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0089 - root_mean_squared_error: 0.0946 - val_loss: 0.0142 - val_root_me

Epoch 32/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0016 - root_mean_squared_error: 0.0396 - val_loss: 0.0476 - val_root_mean_squared_error: 0.2182
Epoch 33/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0016 - root_mean_squared_error: 0.0394 - val_loss: 0.0474 - val_root_mean_squared_error: 0.2177
Epoch 34/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0015 - root_mean_squared_error: 0.0393 - val_loss: 0.0472 - val_root_mean_squared_error: 0.2172
Epoch 35/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0015 - root_mean_squared_error: 0.0391 - val_loss: 0.0469 - val_root_mean_squared_error: 0.2167
Epoch 36/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0015 - root_mean_squared_error: 0.0390 - val_loss: 0.0468 - val_root_mean_squared_error: 0.2162
Epoch 37/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0015 - root_mean_squared_error: 0.0389 - val_loss: 0.0466 

13/13 [==============================] - 0s 25ms/step - loss: 0.0014 - root_mean_squared_error: 0.0376 - val_loss: 0.0430 - val_root_mean_squared_error: 0.2073
Epoch 80/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0014 - root_mean_squared_error: 0.0376 - val_loss: 0.0429 - val_root_mean_squared_error: 0.2072
Epoch 81/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0014 - root_mean_squared_error: 0.0376 - val_loss: 0.0429 - val_root_mean_squared_error: 0.2072
Epoch 82/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0014 - root_mean_squared_error: 0.0376 - val_loss: 0.0429 - val_root_mean_squared_error: 0.2071
Epoch 83/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0014 - root_mean_squared_error: 0.0376 - val_loss: 0.0429 - val_root_mean_squared_error: 0.2071
Epoch 84/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0014 - root_mean_squared_error: 0.0376 - val_loss: 0.0428 - val_root_me

Epoch 9/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0215 - root_mean_squared_error: 0.1465 - val_loss: 0.0520 - val_root_mean_squared_error: 0.2280
Epoch 10/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0202 - root_mean_squared_error: 0.1420 - val_loss: 0.0495 - val_root_mean_squared_error: 0.2225
Epoch 11/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0190 - root_mean_squared_error: 0.1379 - val_loss: 0.0471 - val_root_mean_squared_error: 0.2170
Epoch 12/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0179 - root_mean_squared_error: 0.1339 - val_loss: 0.0450 - val_root_mean_squared_error: 0.2120
Epoch 13/100
13/13 [==============================] - 0s 26ms/step - loss: 0.0170 - root_mean_squared_error: 0.1304 - val_loss: 0.0429 - val_root_mean_squared_error: 0.2072
Epoch 14/100
13/13 [==============================] - 0s 26ms/step - loss: 0.0161 - root_mean_squared_error: 0.1270 - val_loss: 0.0410 -

13/13 [==============================] - 0s 24ms/step - loss: 0.0080 - root_mean_squared_error: 0.0894 - val_loss: 0.0156 - val_root_mean_squared_error: 0.1249
Epoch 57/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0080 - root_mean_squared_error: 0.0894 - val_loss: 0.0155 - val_root_mean_squared_error: 0.1244
Epoch 58/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0080 - root_mean_squared_error: 0.0893 - val_loss: 0.0154 - val_root_mean_squared_error: 0.1239
Epoch 59/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0080 - root_mean_squared_error: 0.0892 - val_loss: 0.0152 - val_root_mean_squared_error: 0.1234
Epoch 60/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0079 - root_mean_squared_error: 0.0891 - val_loss: 0.0151 - val_root_mean_squared_error: 0.1230
Epoch 61/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0079 - root_mean_squared_error: 0.0891 - val_loss: 0.0150 - val_root_me

Epoch 3/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0204 - root_mean_squared_error: 0.1427 - val_loss: 0.0505 - val_root_mean_squared_error: 0.2248
Epoch 4/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0194 - root_mean_squared_error: 0.1393 - val_loss: 0.0485 - val_root_mean_squared_error: 0.2202
Epoch 5/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0185 - root_mean_squared_error: 0.1362 - val_loss: 0.0465 - val_root_mean_squared_error: 0.2157
Epoch 6/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0177 - root_mean_squared_error: 0.1332 - val_loss: 0.0448 - val_root_mean_squared_error: 0.2117
Epoch 7/100
13/13 [==============================] - 0s 26ms/step - loss: 0.0170 - root_mean_squared_error: 0.1305 - val_loss: 0.0431 - val_root_mean_squared_error: 0.2076
Epoch 8/100
13/13 [==============================] - 0s 26ms/step - loss: 0.0164 - root_mean_squared_error: 0.1279 - val_loss: 0.0415 - val_

13/13 [==============================] - 0s 23ms/step - loss: 0.0095 - root_mean_squared_error: 0.0973 - val_loss: 0.0179 - val_root_mean_squared_error: 0.1337
Epoch 51/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0095 - root_mean_squared_error: 0.0972 - val_loss: 0.0177 - val_root_mean_squared_error: 0.1332
Epoch 52/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0094 - root_mean_squared_error: 0.0971 - val_loss: 0.0176 - val_root_mean_squared_error: 0.1326
Epoch 53/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0094 - root_mean_squared_error: 0.0970 - val_loss: 0.0175 - val_root_mean_squared_error: 0.1322
Epoch 54/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0094 - root_mean_squared_error: 0.0970 - val_loss: 0.0173 - val_root_mean_squared_error: 0.1316
Epoch 55/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0094 - root_mean_squared_error: 0.0969 - val_loss: 0.0172 - val_root_me

13/13 [==============================] - 0s 23ms/step - loss: 0.0091 - root_mean_squared_error: 0.0953 - val_loss: 0.0149 - val_root_mean_squared_error: 0.1220
Epoch 98/100
13/13 [==============================] - 0s 22ms/step - loss: 0.0091 - root_mean_squared_error: 0.0952 - val_loss: 0.0148 - val_root_mean_squared_error: 0.1219
Epoch 99/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0091 - root_mean_squared_error: 0.0952 - val_loss: 0.0148 - val_root_mean_squared_error: 0.1217
Epoch 100/100
7/7 [==============================] - 0s 6ms/step - loss: 0.0021 - root_mean_squared_error: 0.0461
Epoch 1/100
13/13 [==============================] - 3s 73ms/step - loss: 0.0048 - root_mean_squared_error: 0.0690 - val_loss: 0.0786 - val_root_mean_squared_error: 0.2803
Epoch 2/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0045 - root_mean_squared_error: 0.0668 - val_loss: 0.0770 - val_root_mean_squared_error: 0.2776
Epoch 3/100
13/13 [=================

Epoch 44/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0017 - root_mean_squared_error: 0.0413 - val_loss: 0.0537 - val_root_mean_squared_error: 0.2318
Epoch 45/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0017 - root_mean_squared_error: 0.0412 - val_loss: 0.0536 - val_root_mean_squared_error: 0.2315
Epoch 46/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0017 - root_mean_squared_error: 0.0412 - val_loss: 0.0534 - val_root_mean_squared_error: 0.2312
Epoch 47/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0017 - root_mean_squared_error: 0.0411 - val_loss: 0.0533 - val_root_mean_squared_error: 0.2309
Epoch 48/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0017 - root_mean_squared_error: 0.0411 - val_loss: 0.0532 - val_root_mean_squared_error: 0.2306
Epoch 49/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0017 - root_mean_squared_error: 0.0411 - val_loss: 0.0531 

13/13 [==============================] - 0s 23ms/step - loss: 0.0017 - root_mean_squared_error: 0.0407 - val_loss: 0.0510 - val_root_mean_squared_error: 0.2259
Epoch 92/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0017 - root_mean_squared_error: 0.0407 - val_loss: 0.0510 - val_root_mean_squared_error: 0.2259
Epoch 93/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0017 - root_mean_squared_error: 0.0407 - val_loss: 0.0510 - val_root_mean_squared_error: 0.2258
Epoch 94/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0017 - root_mean_squared_error: 0.0407 - val_loss: 0.0510 - val_root_mean_squared_error: 0.2258
Epoch 95/100
13/13 [==============================] - 0s 23ms/step - loss: 0.0017 - root_mean_squared_error: 0.0407 - val_loss: 0.0510 - val_root_mean_squared_error: 0.2258
Epoch 96/100
13/13 [==============================] - 0s 24ms/step - loss: 0.0017 - root_mean_squared_error: 0.0407 - val_loss: 0.0510 - val_root_me

In [16]:
mlp_LSTM_model = KerasRegressor(multiple_layer_LSTM_model_plus) 
params = dict(optimizer=['sgd', 'adam'], epochs=[50, 100], batch_size=[16, 32], activation=['relu','tanh'],
             learning_rate=[0.01, 0.001, 0.0005])
random_search = RandomizedSearchCV(mlp_LSTM_model, params, cv=3)
mlp_LSTM_results = random_search.fit(x_train, y_train, validation_data=(x_valid, y_valid),
                                          callbacks=[EarlyStopping(patience=3)])
print(f"Best: {mlp_LSTM_results.best_score_} using {mlp_LSTM_results.best_params_}")

<ipython-input-16-479eaed71c97>:1: DeprecationWarning: KerasRegressor is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  mlp_LSTM_model = KerasRegressor(multiple_layer_LSTM_model_plus)


Epoch 1/50
7/7 [==============================] - 3s 113ms/step - loss: 0.0125 - root_mean_squared_error: 0.1119 - val_loss: 0.0015 - val_root_mean_squared_error: 0.0387
Epoch 2/50
7/7 [==============================] - 0s 29ms/step - loss: 0.0033 - root_mean_squared_error: 0.0571 - val_loss: 0.0076 - val_root_mean_squared_error: 0.0873
Epoch 3/50
7/7 [==============================] - 0s 30ms/step - loss: 0.0025 - root_mean_squared_error: 0.0500 - val_loss: 0.0031 - val_root_mean_squared_error: 0.0558
Epoch 4/50
7/7 [==============================] - 0s 31ms/step - loss: 0.0012 - root_mean_squared_error: 0.0341 - val_loss: 2.7936e-04 - val_root_mean_squared_error: 0.0167
Epoch 5/50
7/7 [==============================] - 0s 31ms/step - loss: 8.8166e-04 - root_mean_squared_error: 0.0297 - val_loss: 0.0017 - val_root_mean_squared_error: 0.0413
Epoch 6/50
7/7 [==============================] - 0s 31ms/step - loss: 6.5059e-04 - root_mean_squared_error: 0.0255 - val_loss: 4.9701e-04 - val_r

7/7 [==============================] - 3s 115ms/step - loss: 0.0179 - root_mean_squared_error: 0.1337 - val_loss: 0.0343 - val_root_mean_squared_error: 0.1852
Epoch 2/50
7/7 [==============================] - 0s 28ms/step - loss: 0.0118 - root_mean_squared_error: 0.1084 - val_loss: 0.0245 - val_root_mean_squared_error: 0.1567
Epoch 3/50
7/7 [==============================] - 0s 30ms/step - loss: 0.0097 - root_mean_squared_error: 0.0983 - val_loss: 0.0191 - val_root_mean_squared_error: 0.1382
Epoch 4/50
7/7 [==============================] - 0s 30ms/step - loss: 0.0088 - root_mean_squared_error: 0.0937 - val_loss: 0.0162 - val_root_mean_squared_error: 0.1271
Epoch 5/50
7/7 [==============================] - 0s 33ms/step - loss: 0.0084 - root_mean_squared_error: 0.0917 - val_loss: 0.0144 - val_root_mean_squared_error: 0.1198
Epoch 6/50
7/7 [==============================] - 0s 31ms/step - loss: 0.0082 - root_mean_squared_error: 0.0907 - val_loss: 0.0137 - val_root_mean_squared_error: 0.1

7/7 [==============================] - 0s 32ms/step - loss: 0.0083 - root_mean_squared_error: 0.0911 - val_loss: 0.0218 - val_root_mean_squared_error: 0.1476
Epoch 13/50
7/7 [==============================] - 0s 31ms/step - loss: 0.0078 - root_mean_squared_error: 0.0884 - val_loss: 0.0207 - val_root_mean_squared_error: 0.1439
Epoch 14/50
7/7 [==============================] - 0s 33ms/step - loss: 0.0074 - root_mean_squared_error: 0.0860 - val_loss: 0.0197 - val_root_mean_squared_error: 0.1404
Epoch 15/50
7/7 [==============================] - 0s 34ms/step - loss: 0.0070 - root_mean_squared_error: 0.0838 - val_loss: 0.0188 - val_root_mean_squared_error: 0.1372
Epoch 16/50
7/7 [==============================] - 0s 34ms/step - loss: 0.0067 - root_mean_squared_error: 0.0819 - val_loss: 0.0180 - val_root_mean_squared_error: 0.1342
Epoch 17/50
7/7 [==============================] - 0s 35ms/step - loss: 0.0064 - root_mean_squared_error: 0.0802 - val_loss: 0.0172 - val_root_mean_squared_error:

Epoch 10/50
7/7 [==============================] - 0s 31ms/step - loss: 0.0144 - root_mean_squared_error: 0.1200 - val_loss: 0.0365 - val_root_mean_squared_error: 0.1910
Epoch 11/50
7/7 [==============================] - 0s 30ms/step - loss: 0.0139 - root_mean_squared_error: 0.1179 - val_loss: 0.0352 - val_root_mean_squared_error: 0.1877
Epoch 12/50
7/7 [==============================] - 0s 31ms/step - loss: 0.0135 - root_mean_squared_error: 0.1160 - val_loss: 0.0341 - val_root_mean_squared_error: 0.1845
Epoch 13/50
7/7 [==============================] - 0s 35ms/step - loss: 0.0131 - root_mean_squared_error: 0.1143 - val_loss: 0.0328 - val_root_mean_squared_error: 0.1812
Epoch 14/50
7/7 [==============================] - 0s 34ms/step - loss: 0.0127 - root_mean_squared_error: 0.1126 - val_loss: 0.0317 - val_root_mean_squared_error: 0.1781
Epoch 15/50
7/7 [==============================] - 0s 33ms/step - loss: 0.0123 - root_mean_squared_error: 0.1111 - val_loss: 0.0307 - val_root_mean_sq

Epoch 8/50
7/7 [==============================] - 0s 30ms/step - loss: 0.0036 - root_mean_squared_error: 0.0597 - val_loss: 0.0775 - val_root_mean_squared_error: 0.2785
Epoch 9/50
7/7 [==============================] - 0s 30ms/step - loss: 0.0034 - root_mean_squared_error: 0.0584 - val_loss: 0.0765 - val_root_mean_squared_error: 0.2765
Epoch 10/50
7/7 [==============================] - 0s 30ms/step - loss: 0.0033 - root_mean_squared_error: 0.0571 - val_loss: 0.0754 - val_root_mean_squared_error: 0.2746
Epoch 11/50
7/7 [==============================] - 0s 29ms/step - loss: 0.0031 - root_mean_squared_error: 0.0559 - val_loss: 0.0745 - val_root_mean_squared_error: 0.2729
Epoch 12/50
7/7 [==============================] - 0s 30ms/step - loss: 0.0030 - root_mean_squared_error: 0.0548 - val_loss: 0.0735 - val_root_mean_squared_error: 0.2712
Epoch 13/50
7/7 [==============================] - 0s 30ms/step - loss: 0.0029 - root_mean_squared_error: 0.0538 - val_loss: 0.0727 - val_root_mean_squa

7/7 [==============================] - 0s 6ms/step - loss: 1.3194e-04 - root_mean_squared_error: 0.0115
Epoch 1/100
13/13 [==============================] - 3s 68ms/step - loss: 0.0070 - root_mean_squared_error: 0.0834 - val_loss: 0.0059 - val_root_mean_squared_error: 0.0766
Epoch 2/100
13/13 [==============================] - 0s 22ms/step - loss: 0.0013 - root_mean_squared_error: 0.0363 - val_loss: 8.2399e-04 - val_root_mean_squared_error: 0.0287
Epoch 3/100
13/13 [==============================] - 0s 23ms/step - loss: 6.0990e-04 - root_mean_squared_error: 0.0247 - val_loss: 5.4139e-04 - val_root_mean_squared_error: 0.0233
Epoch 4/100
13/13 [==============================] - 0s 23ms/step - loss: 4.3205e-04 - root_mean_squared_error: 0.0208 - val_loss: 5.5614e-04 - val_root_mean_squared_error: 0.0236
Epoch 5/100
13/13 [==============================] - 0s 24ms/step - loss: 3.8350e-04 - root_mean_squared_error: 0.0196 - val_loss: 4.9988e-04 - val_root_mean_squared_error: 0.0224
Epoch 6/

7/7 [==============================] - 0s 7ms/step - loss: 0.0016 - root_mean_squared_error: 0.0402
Epoch 1/100
13/13 [==============================] - 3s 70ms/step - loss: 0.0290 - root_mean_squared_error: 0.1702 - val_loss: 0.0602 - val_root_mean_squared_error: 0.2454
Epoch 2/100
13/13 [==============================] - 0s 26ms/step - loss: 0.0246 - root_mean_squared_error: 0.1569 - val_loss: 0.0533 - val_root_mean_squared_error: 0.2309
Epoch 3/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0211 - root_mean_squared_error: 0.1452 - val_loss: 0.0475 - val_root_mean_squared_error: 0.2179
Epoch 4/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0182 - root_mean_squared_error: 0.1351 - val_loss: 0.0426 - val_root_mean_squared_error: 0.2064
Epoch 5/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0160 - root_mean_squared_error: 0.1264 - val_loss: 0.0385 - val_root_mean_squared_error: 0.1961
Epoch 6/100
13/13 [=====================

Epoch 47/100
13/13 [==============================] - 0s 30ms/step - loss: 0.0062 - root_mean_squared_error: 0.0788 - val_loss: 0.0108 - val_root_mean_squared_error: 0.1041
Epoch 48/100
13/13 [==============================] - 0s 28ms/step - loss: 0.0062 - root_mean_squared_error: 0.0787 - val_loss: 0.0108 - val_root_mean_squared_error: 0.1041
Epoch 49/100
13/13 [==============================] - 0s 28ms/step - loss: 0.0062 - root_mean_squared_error: 0.0787 - val_loss: 0.0108 - val_root_mean_squared_error: 0.1040
Epoch 50/100
13/13 [==============================] - 0s 30ms/step - loss: 0.0062 - root_mean_squared_error: 0.0787 - val_loss: 0.0108 - val_root_mean_squared_error: 0.1039
Epoch 51/100
13/13 [==============================] - 0s 30ms/step - loss: 0.0062 - root_mean_squared_error: 0.0786 - val_loss: 0.0108 - val_root_mean_squared_error: 0.1038
Epoch 52/100
13/13 [==============================] - 0s 35ms/step - loss: 0.0062 - root_mean_squared_error: 0.0786 - val_loss: 0.0107 

Epoch 32/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0118 - root_mean_squared_error: 0.1087 - val_loss: 0.0221 - val_root_mean_squared_error: 0.1486
Epoch 33/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0118 - root_mean_squared_error: 0.1086 - val_loss: 0.0218 - val_root_mean_squared_error: 0.1476
Epoch 34/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0118 - root_mean_squared_error: 0.1084 - val_loss: 0.0215 - val_root_mean_squared_error: 0.1466
Epoch 35/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0117 - root_mean_squared_error: 0.1083 - val_loss: 0.0213 - val_root_mean_squared_error: 0.1458
Epoch 36/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0117 - root_mean_squared_error: 0.1081 - val_loss: 0.0210 - val_root_mean_squared_error: 0.1449
Epoch 37/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0117 - root_mean_squared_error: 0.1080 - val_loss: 0.0208 

13/13 [==============================] - 0s 29ms/step - loss: 0.0111 - root_mean_squared_error: 0.1054 - val_loss: 0.0178 - val_root_mean_squared_error: 0.1335
Epoch 80/100
13/13 [==============================] - 0s 29ms/step - loss: 0.0111 - root_mean_squared_error: 0.1054 - val_loss: 0.0178 - val_root_mean_squared_error: 0.1333
Epoch 81/100
13/13 [==============================] - 0s 29ms/step - loss: 0.0111 - root_mean_squared_error: 0.1053 - val_loss: 0.0177 - val_root_mean_squared_error: 0.1332
Epoch 82/100
13/13 [==============================] - 0s 32ms/step - loss: 0.0111 - root_mean_squared_error: 0.1052 - val_loss: 0.0177 - val_root_mean_squared_error: 0.1331
Epoch 83/100
13/13 [==============================] - 0s 30ms/step - loss: 0.0111 - root_mean_squared_error: 0.1052 - val_loss: 0.0177 - val_root_mean_squared_error: 0.1330
Epoch 84/100
13/13 [==============================] - 0s 28ms/step - loss: 0.0111 - root_mean_squared_error: 0.1051 - val_loss: 0.0176 - val_root_me

Epoch 26/100
13/13 [==============================] - 0s 30ms/step - loss: 0.0015 - root_mean_squared_error: 0.0382 - val_loss: 0.0459 - val_root_mean_squared_error: 0.2142
Epoch 27/100
13/13 [==============================] - 0s 29ms/step - loss: 0.0015 - root_mean_squared_error: 0.0381 - val_loss: 0.0457 - val_root_mean_squared_error: 0.2138
Epoch 28/100
13/13 [==============================] - 0s 28ms/step - loss: 0.0015 - root_mean_squared_error: 0.0381 - val_loss: 0.0455 - val_root_mean_squared_error: 0.2134
Epoch 29/100
13/13 [==============================] - 0s 28ms/step - loss: 0.0014 - root_mean_squared_error: 0.0380 - val_loss: 0.0454 - val_root_mean_squared_error: 0.2130
Epoch 30/100
13/13 [==============================] - 0s 28ms/step - loss: 0.0014 - root_mean_squared_error: 0.0380 - val_loss: 0.0452 - val_root_mean_squared_error: 0.2127
Epoch 31/100
13/13 [==============================] - 0s 29ms/step - loss: 0.0014 - root_mean_squared_error: 0.0380 - val_loss: 0.0451 

13/13 [==============================] - 0s 27ms/step - loss: 0.0014 - root_mean_squared_error: 0.0377 - val_loss: 0.0436 - val_root_mean_squared_error: 0.2088
Epoch 74/100
13/13 [==============================] - 0s 28ms/step - loss: 0.0014 - root_mean_squared_error: 0.0378 - val_loss: 0.0436 - val_root_mean_squared_error: 0.2088
Epoch 75/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0014 - root_mean_squared_error: 0.0378 - val_loss: 0.0436 - val_root_mean_squared_error: 0.2088
Epoch 76/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0014 - root_mean_squared_error: 0.0377 - val_loss: 0.0436 - val_root_mean_squared_error: 0.2088
Epoch 77/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0014 - root_mean_squared_error: 0.0377 - val_loss: 0.0436 - val_root_mean_squared_error: 0.2088
Epoch 78/100
13/13 [==============================] - 0s 26ms/step - loss: 0.0014 - root_mean_squared_error: 0.0377 - val_loss: 0.0436 - val_root_me

13/13 [==============================] - 0s 26ms/step - loss: 0.0035 - root_mean_squared_error: 0.0593 - val_loss: 0.0069 - val_root_mean_squared_error: 0.0830
Epoch 25/100
13/13 [==============================] - 0s 26ms/step - loss: 0.0034 - root_mean_squared_error: 0.0585 - val_loss: 0.0059 - val_root_mean_squared_error: 0.0768
Epoch 26/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0033 - root_mean_squared_error: 0.0577 - val_loss: 0.0055 - val_root_mean_squared_error: 0.0740
Epoch 27/100
13/13 [==============================] - 0s 27ms/step - loss: 0.0032 - root_mean_squared_error: 0.0570 - val_loss: 0.0057 - val_root_mean_squared_error: 0.0755
Epoch 28/100
13/13 [==============================] - 0s 25ms/step - loss: 0.0032 - root_mean_squared_error: 0.0563 - val_loss: 0.0059 - val_root_mean_squared_error: 0.0768
Epoch 29/100
7/7 [==============================] - 0s 6ms/step - loss: 5.8695e-04 - root_mean_squared_error: 0.0242
Epoch 1/100
13/13 [============

Epoch 4/100
7/7 [==============================] - 0s 33ms/step - loss: 1.9122e-04 - root_mean_squared_error: 0.0138 - val_loss: 4.4191e-04 - val_root_mean_squared_error: 0.0210
Epoch 5/100
7/7 [==============================] - 0s 32ms/step - loss: 1.1293e-04 - root_mean_squared_error: 0.0106 - val_loss: 8.3292e-04 - val_root_mean_squared_error: 0.0289
Epoch 6/100
7/7 [==============================] - 0s 33ms/step - loss: 7.8854e-05 - root_mean_squared_error: 0.0089 - val_loss: 3.1886e-04 - val_root_mean_squared_error: 0.0179
Epoch 7/100
7/7 [==============================] - 0s 32ms/step - loss: 6.6443e-05 - root_mean_squared_error: 0.0082 - val_loss: 1.9386e-04 - val_root_mean_squared_error: 0.0139
Epoch 8/100
7/7 [==============================] - 0s 32ms/step - loss: 5.3900e-05 - root_mean_squared_error: 0.0073 - val_loss: 7.7486e-04 - val_root_mean_squared_error: 0.0278
Epoch 9/100
7/7 [==============================] - 0s 32ms/step - loss: 4.5716e-05 - root_mean_squared_error: 

Epoch 12/100
7/7 [==============================] - 0s 35ms/step - loss: 0.0051 - root_mean_squared_error: 0.0715 - val_loss: 0.0093 - val_root_mean_squared_error: 0.0963
Epoch 13/100
4/4 [==============================] - 0s 8ms/step - loss: 0.0135 - root_mean_squared_error: 0.1163
Epoch 1/100
7/7 [==============================] - 3s 158ms/step - loss: 0.0209 - root_mean_squared_error: 0.1446 - val_loss: 0.0362 - val_root_mean_squared_error: 0.1904
Epoch 2/100
7/7 [==============================] - 0s 32ms/step - loss: 0.0118 - root_mean_squared_error: 0.1086 - val_loss: 0.0242 - val_root_mean_squared_error: 0.1556
Epoch 3/100
7/7 [==============================] - 0s 34ms/step - loss: 0.0091 - root_mean_squared_error: 0.0955 - val_loss: 0.0192 - val_root_mean_squared_error: 0.1387
Epoch 4/100
7/7 [==============================] - 0s 34ms/step - loss: 0.0081 - root_mean_squared_error: 0.0901 - val_loss: 0.0153 - val_root_mean_squared_error: 0.1239
Epoch 5/100
7/7 [==================

### Best Model 

In [17]:
rnn = round(abs(rnn_model_results.best_score_), 6)
sl = round(abs(sl_LSTM_results.best_score_), 6)
ml = round(abs(ml_LSTM_results.best_score_), 6)
mlp = round(abs(mlp_LSTM_results.best_score_), 6)

print(f"The Root Mean Squared Error for single layer RNN model is {rnn}")
print(f"The Root Mean Squared Error for single layer LSTM model is {sl}")
print(f"The Root Mean Squared Error for two layers LSTM model is {ml}")
print(f"The Root Mean Squared Error for two layers LSTM model with a dense layer is {mlp}")

The Root Mean Squared Error for single layer RNN model is 0.000213
The Root Mean Squared Error for single layer LSTM model is 0.000146
The Root Mean Squared Error for two layers LSTM model is 0.000433
The Root Mean Squared Error for two layers LSTM model with a dense layer is 0.000442


In [18]:
sl_LSTM_results.best_params_

{'optimizer': 'adam',
 'learning_rate': 0.01,
 'epochs': 100,
 'batch_size': 32,
 'activation': 'relu'}

In [21]:
# Configure the neural network model
model = Sequential()

neurons = sequence_length 

# Model with sequence_length Neurons 
# inputshape = sequence_length Timestamps 
model.add(LSTM(neurons, activation="relu", return_sequences=False, input_shape=(x_train.shape[1], 1))) 
model.add(Dense(1)) 

optimizer = Adam(learning_rate=0.01) 
# Compile the model 
model.compile(optimizer='adam', loss='mean_squared_error') 

# Training the model 
model.fit(x_train, y_train, batch_size=32, epochs=100) 

Epoch 1/100
10/10 [==============================] - 1s 13ms/step - loss: 0.0092
Epoch 2/100
10/10 [==============================] - 0s 13ms/step - loss: 0.0026
Epoch 3/100
10/10 [==============================] - 0s 13ms/step - loss: 0.0017
Epoch 4/100
10/10 [==============================] - 0s 14ms/step - loss: 8.5511e-04
Epoch 5/100
10/10 [==============================] - 0s 13ms/step - loss: 4.1130e-04
Epoch 6/100
10/10 [==============================] - 0s 13ms/step - loss: 3.2282e-04
Epoch 7/100
10/10 [==============================] - 0s 13ms/step - loss: 2.7647e-04
Epoch 8/100
10/10 [==============================] - 0s 13ms/step - loss: 2.6408e-04
Epoch 9/100
10/10 [==============================] - 0s 13ms/step - loss: 2.3805e-04
Epoch 10/100
10/10 [==============================] - 0s 13ms/step - loss: 2.3008e-04
Epoch 11/100
10/10 [==============================] - 0s 13ms/step - loss: 2.1493e-04
Epoch 12/100
10/10 [==============================] - 0s 12ms/step - loss: 

10/10 [==============================] - 0s 14ms/step - loss: 5.6739e-05
Epoch 97/100
10/10 [==============================] - 0s 14ms/step - loss: 6.0651e-05
Epoch 98/100
10/10 [==============================] - 0s 13ms/step - loss: 5.8581e-05
Epoch 99/100
10/10 [==============================] - 0s 13ms/step - loss: 6.3481e-05
Epoch 100/100
10/10 [==============================] - 0s 14ms/step - loss: 5.7750e-05


In [22]:
model.save('saved_model/uni_model')

INFO:tensorflow:Assets written to: saved_model/uni_model\assets


INFO:tensorflow:Assets written to: saved_model/uni_model\assets
